{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ff8b5bf5-4f47-4996-8cd9-67b2a4d765f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Seq, SeqIO, SeqRecord\n",
    "from IPython.display import display\n",
    "from plotnine import *\n",
    "from scipy.optimize import curve_fit\n",
    "from tools import *\n",
    "import math\n",
    "import os\n",
    "import pandas as pd\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a279853a-3469-4f4b-9815-751da9a67c02",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_dir = \"input\"\n",
    "output_dir = \"output\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4b43d849-d7b1-448f-95d2-f40b0c70eb84",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> Assembly indexing\n",
      "14 / 14 runs were skipped.\n",
      "Assembly number: 14\n"
     ]
    }
   ],
   "source": [
    "# Assembly indexing\n",
    "print(\">\", \"Assembly indexing\")\n",
    "asm_i_dir = input_dir + \"/asm\"\n",
    "asm_o_dir = output_dir + \"/asm\"\n",
    "if not os.path.exists(asm_o_dir):\n",
    "    os.makedirs(asm_o_dir)\n",
    "asms = []\n",
    "skipped = 0\n",
    "fls = sorted([i for i in os.listdir(asm_i_dir) if i.endswith(\".fa\") or fl.endswith(\".fasta\")])\n",
    "for fl in fls:\n",
    "    asm = os.path.splitext(fl)[0]\n",
    "    if asm in asms:\n",
    "        print(\"Assembly name duplicated: {}.\".format(asm), file=sys.stderr)\n",
    "        break\n",
    "    else:\n",
    "        asms.append(asm)\n",
    "    #print(\">\", asm)\n",
    "    asm_file = asm_o_dir + \"/\" + asm + \".fa\"\n",
    "    if not os.path.exists(asm_file):\n",
    "        os.symlink(os.path.abspath(asm_i_dir + \"/\" + fl), asm_file)\n",
    "        os.system(\"samtools faidx {}\".format(asm_file))\n",
    "    else:\n",
    "        skipped += 1\n",
    "        #print(\"{} exists, skipped.\".format(asm_file))\n",
    "if skipped > 0:\n",
    "    print(\"{} / {} runs were skipped.\".format(skipped, len(fls)))\n",
    "print(\"Assembly number: {}\".format(len([i for i in os.listdir(asm_o_dir) if i.endswith(\".fa\")])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a8e355eb-47ee-4607-9e93-e8ff69f13576",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> CRISPR identification\n",
      "28 / 28 runs were skipped.\n"
     ]
    }
   ],
   "source": [
    "# CRISPR identification\n",
    "print(\">\", \"CRISPR identification\")\n",
    "cri_dir = output_dir + \"/crispr\"\n",
    "if not os.path.exists(cri_dir):\n",
    "    os.makedirs(cri_dir)\n",
    "skipped = 0\n",
    "fls = sorted([i for i in os.listdir(asm_o_dir) if i.endswith(\".fa\")])\n",
    "for fl in fls:\n",
    "    asm = os.path.splitext(fl)[0]\n",
    "    #print(\">\", asm)\n",
    "    asm_file = os.path.join(asm_o_dir, fl)\n",
    "    # PilerCR\n",
    "    if os.path.exists(\"{}/{}_pilerCR.out\".format(cri_dir, asm)):\n",
    "        skipped += 1\n",
    "        #print(\"{}/{}_pilerCR.out exists, skipped.\".format(cri_dir, asm))\n",
    "    else:\n",
    "        pcmd = \"pilercr -minid 0.85 -mincons 0.8 -noinfo -in {} -out {}/{}_pilerCR.out 2>{}/{}_pilerCR.err\".format(asm_file, cri_dir, asm, cri_dir, asm)\n",
    "        retCode1 = os.system(pcmd)\n",
    "        if retCode1 != 0:\n",
    "            print(str(retCode1) + \"  \" + pcmd)\n",
    "            break\n",
    "    # minced\n",
    "    if os.path.exists(\"{}/{}_minCED.out\".format(cri_dir, asm)):\n",
    "        skipped += 1\n",
    "        #print(\"{}/{}_minCED.out exists, skipped.\".format(cri_dir, asm))\n",
    "    else:\n",
    "        mcmd = \"minced -minRL 16 -maxRL 64 -minSL 8 -maxSL 64 -searchWL 6 {} {}/{}_minCED.out 2>{}/{}_minCED.err\".format(asm_file, cri_dir, asm, cri_dir, asm)\n",
    "        retCode2 = os.system(mcmd)\n",
    "        if retCode2 != 0:\n",
    "            print(str(retCode2) + \"  \" + mcmd)\n",
    "            break\n",
    "if skipped > 0:\n",
    "    print(\"{} / {} runs were skipped.\".format(skipped, len(fls) * 2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c53b25a3-807f-4cb6-80c6-04a7c04a2d7a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> CRISPR res extraction\n",
      "criRes (3789, 12)\n",
      "All the sequences have been validated successfully.\n",
      "conRes (237, 7)\n",
      "12 / 14 assemblies contain CRISPRs.\n"
     ]
    }
   ],
   "source": [
    "# CRISPR res extraction\n",
    "print(\">\", \"CRISPR res extraction\")\n",
    "all_dir = output_dir + \"/all\"\n",
    "if not os.path.exists(all_dir):\n",
    "    os.makedirs(all_dir)\n",
    "criRes = pd.DataFrame()\n",
    "conRes = pd.DataFrame()\n",
    "for fl in sorted(os.listdir(cri_dir)):\n",
    "    #asm = os.path.splitext(fl)[0]\n",
    "    #print(\">\", asm)\n",
    "    if \"pilerCR.out\" in fl:\n",
    "        crip, copp = readPIL(os.path.join(cri_dir, fl))\n",
    "        criRes = pd.concat([criRes, crip], ignore_index=True)\n",
    "        conRes = pd.concat([conRes, copp], ignore_index=True)\n",
    "    elif \"minCED.out\" in fl:\n",
    "        crip, copp = readMIN(os.path.join(cri_dir, fl))\n",
    "        criRes = pd.concat([criRes, crip], ignore_index=True)\n",
    "        conRes = pd.concat([conRes, copp], ignore_index=True)\n",
    "    elif fl.endswith(\".err\"):\n",
    "        continue\n",
    "    else:\n",
    "        print(\"Unrecognized CRISPR identification file: {}\".format(os.path.join(cri_dir, fl)), file=sys.stderr)\n",
    "print(\"criRes\", criRes.shape)\n",
    "validateSeq(criRes, genomeDir=asm_o_dir)\n",
    "print(\"conRes\", conRes.shape)\n",
    "print(\"{} / {} assemblies contain CRISPRs.\".format(len(criRes[\"asm\"].unique()), len(asms)))\n",
    "criNrs = criRes.pivot_table(index=[\"cid\", \"asm\"], columns=\"component\", values=\"id\", aggfunc=len).reset_index()\n",
    "criPos = criRes.groupby([\"cid\", \"asm\"]).agg({\"contig\": lambda x: x.unique()[0], \"start\": min, \"end\": max}).reset_index()\n",
    "criPos[\"length\"] = criPos[\"end\"] - criPos[\"start\"]\n",
    "criMet = pd.merge(criNrs, criPos.drop(columns=\"asm\"), on=\"cid\")[[\"contig\", \"start\", \"end\", \"cid\", \"repeat\", \"spacer\", \"length\", \"asm\"]]\n",
    "repRes = criRes[criRes[\"component\"] == \"repeat\"]\n",
    "spaRes = criRes[criRes[\"component\"] == \"spacer\"]\n",
    "# write txt\n",
    "repRes.to_csv(all_dir + \"/repeat.txt\", header=True, sep=\"\\t\", quoting=csv.QUOTE_NONE, index=False)\n",
    "spaRes.to_csv(all_dir + \"/spacer.txt\", header=True, sep=\"\\t\", quoting=csv.QUOTE_NONE, index=False)\n",
    "conRes.to_csv(all_dir + \"/conRep.txt\", header=True, sep=\"\\t\", quoting=csv.QUOTE_NONE, index=False)\n",
    "criMet.to_csv(all_dir + \"/criMet.txt\", header=True, sep=\"\\t\", quoting=csv.QUOTE_NONE, index=False)\n",
    "# write fa\n",
    "recs = [SeqRecord.SeqRecord(Seq.Seq(row[\"seq\"]), id=row[\"id\"], description=\"\") for i, row in repRes.iterrows()]\n",
    "SeqIO.write(recs, all_dir + \"/repeat.fa\", \"fasta\")\n",
    "recs = [SeqRecord.SeqRecord(Seq.Seq(row[\"seq\"]), id=row[\"id\"], description=\"\") for i, row in spaRes.iterrows()]\n",
    "SeqIO.write(recs, all_dir + \"/spacer.fa\", \"fasta\")\n",
    "recs = [SeqRecord.SeqRecord(Seq.Seq(row[\"seq\"]), id=row[\"id\"], description=\"\") for i, row in conRes.iterrows()]\n",
    "SeqIO.write(recs, all_dir + \"/conRep.fa\", \"fasta\")\n",
    "# write fa by asm\n",
    "con_dir = output_dir + \"/conRep\"\n",
    "if not os.path.exists(con_dir):\n",
    "    os.makedirs(con_dir)\n",
    "for asm, df in conRes.groupby(\"asm\"):\n",
    "    recs = [SeqRecord.SeqRecord(Seq.Seq(row[\"seq\"]), id=row[\"id\"], description=\"\") for i, row in df.iterrows()]\n",
    "    SeqIO.write(recs, con_dir + \"/\" + asm + \".fa\", \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "700163d7-24dc-44ff-8a45-d4f489c1adea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# stat\n",
    "# repeats or spacers for each asm\n",
    "stat = criRes.pivot_table(index=[\"asm\", \"tool\"], columns=\"component\", values=\"seq\",\n",
    "                          aggfunc=lambda x: (len(x), str(min([len(i) for i in x])) + \"-\" + str(max([len(i) for i in x])))).reset_index()\n",
    "stat.columns.name = None\n",
    "stat.to_csv(all_dir + \"/crispr_stat.txt\", header=True, sep=\"\\t\", quoting=csv.QUOTE_NONE, index=False)\n",
    "# consensus repeats for each asm\n",
    "stat_con = conRes.pivot_table(index=[\"asm\", \"tool\"], values=\"cid\", aggfunc=len).reset_index()\n",
    "stat_con.to_csv(all_dir + \"/conRep_stat.txt\", header=True, sep=\"\\t\", quoting=csv.QUOTE_NONE, index=False)\n",
    "# CRISPR number ~ con rep number\n",
    "stat_arr = criRes.pivot_table(index=[\"asm\", \"tool\"], values=\"cid\", aggfunc=lambda x: len(x.unique())).reset_index()\n",
    "assert all(stat_con[\"cid\"] == stat_arr[\"cid\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c2c307fe-3ee0-4e7a-b6da-1316f1d67b72",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> Blast DB construction\n",
      "14 / 14 runs were skipped.\n"
     ]
    }
   ],
   "source": [
    "# Blast DB construction\n",
    "print(\">\", \"Blast DB construction\")\n",
    "skipped = 0\n",
    "fls = sorted([i for i in os.listdir(asm_o_dir) if i.endswith(\".fa\")])\n",
    "for fl in fls:\n",
    "    asm = os.path.splitext(fl)[0]\n",
    "    #print(\">\", asm)\n",
    "    asm_file = os.path.join(asm_o_dir, fl)\n",
    "    if not os.path.exists(asm_file + \".ndb\"):\n",
    "        os.system(\"makeblastdb -in {} -parse_seqids -dbtype nucl > {}\".format(asm_file, asm_file + \".log\"))\n",
    "    else:\n",
    "        skipped += 1\n",
    "        #print(\"{} exists, skipped.\".format(asm_file + \".ndb\"))\n",
    "if skipped > 0:\n",
    "    print(\"{} / {} runs were skipped.\".format(skipped, len(fls)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "aca96c47-908c-4948-8253-f314001ac03e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> Consensus repeat reordering\n"
     ]
    }
   ],
   "source": [
    "# Con repeat reordering\n",
    "print(\">\", \"Consensus repeat reordering\")\n",
    "conReo_dir = output_dir + \"/conReo\"\n",
    "if not os.path.exists(conReo_dir):\n",
    "    os.makedirs(conReo_dir)\n",
    "random.seed(1)\n",
    "for asm, df in conRes.groupby(\"asm\"):\n",
    "    recs = []\n",
    "    for _, row in df.iterrows():\n",
    "        for i in range(100):\n",
    "            bases = list(row[\"seq\"])\n",
    "            random.shuffle(bases)\n",
    "            reo = \"\".join(bases)\n",
    "            recs.append(SeqRecord.SeqRecord(Seq.Seq(reo), id=row[\"id\"] + \"_\" + str(i + 1), description=\"\"))\n",
    "    SeqIO.write(recs, conReo_dir + \"/\" + asm + \".fa\", \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e2cc252a-992a-45b6-90b7-c4e0b5a5353a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> Blast for consensus repeats (reordered)\n",
      "12 / 12 runs were skipped.\n"
     ]
    }
   ],
   "source": [
    "# Blast for consensus repeats (reordered)\n",
    "print(\">\", \"Blast for consensus repeats (reordered)\")\n",
    "blcReo_dir = output_dir + \"/blastConReo\"\n",
    "if not os.path.exists(blcReo_dir):\n",
    "    os.makedirs(blcReo_dir)\n",
    "skipped = 0\n",
    "fls = sorted(os.listdir(conReo_dir))\n",
    "for fl in fls:\n",
    "    asm = os.path.splitext(fl)[0]\n",
    "    #print(\">\", asm)\n",
    "    asm_file = os.path.join(asm_o_dir, fl)\n",
    "    xml_file = os.path.join(blcReo_dir, asm + \".xml\")\n",
    "    if not os.path.exists(xml_file):\n",
    "        os.system(\"blastn -query {} -db {} -task blastn-short -dust no -outfmt 5 -num_threads 12 -out {}\".format(os.path.join(conReo_dir, fl), asm_file, xml_file))\n",
    "    else:\n",
    "        skipped += 1\n",
    "        #print(\"{} exists, skipped.\".format(xml_file))\n",
    "if skipped > 0:\n",
    "    print(\"{} / {} runs were skipped.\".format(skipped, len(fls)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b6f7c07a-c475-4f20-ad93-ebe50d2afcce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> Blast res extraction for consensus repeats (reordered)\n",
      "output/all/blastConReo.txt exists, read it directly.\n",
      "(906605, 20)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>asm</th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>pident</th>\n",
       "      <th>length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gap</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>qlen</th>\n",
       "      <th>qcov</th>\n",
       "      <th>strand</th>\n",
       "      <th>qseq</th>\n",
       "      <th>sseq</th>\n",
       "      <th>alm</th>\n",
       "      <th>tool</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_1</td>\n",
       "      <td>LTBA01000010.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>20</td>\n",
       "      <td>10686</td>\n",
       "      <td>10674</td>\n",
       "      <td>0.645306</td>\n",
       "      <td>26.2635</td>\n",
       "      <td>30</td>\n",
       "      <td>0.433</td>\n",
       "      <td>-</td>\n",
       "      <td>TGTTATGATAGGA</td>\n",
       "      <td>TGTTATGATAGGA</td>\n",
       "      <td>|||||||||||||</td>\n",
       "      <td>minCED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_1</td>\n",
       "      <td>LTBA01000010.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>21</td>\n",
       "      <td>52349</td>\n",
       "      <td>52338</td>\n",
       "      <td>2.549850</td>\n",
       "      <td>24.2811</td>\n",
       "      <td>30</td>\n",
       "      <td>0.400</td>\n",
       "      <td>-</td>\n",
       "      <td>TTATGATAGGAG</td>\n",
       "      <td>TTATGATAGGAG</td>\n",
       "      <td>||||||||||||</td>\n",
       "      <td>minCED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_1</td>\n",
       "      <td>LTBA01000001.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>22</td>\n",
       "      <td>267085</td>\n",
       "      <td>267097</td>\n",
       "      <td>0.645306</td>\n",
       "      <td>26.2635</td>\n",
       "      <td>30</td>\n",
       "      <td>0.433</td>\n",
       "      <td>+</td>\n",
       "      <td>TTATGATAGGAGG</td>\n",
       "      <td>TTATGATAGGAGG</td>\n",
       "      <td>|||||||||||||</td>\n",
       "      <td>minCED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_1</td>\n",
       "      <td>LTBA01000001.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>24</td>\n",
       "      <td>344404</td>\n",
       "      <td>344392</td>\n",
       "      <td>0.645306</td>\n",
       "      <td>26.2635</td>\n",
       "      <td>30</td>\n",
       "      <td>0.433</td>\n",
       "      <td>-</td>\n",
       "      <td>ATGATAGGAGGAT</td>\n",
       "      <td>ATGATAGGAGGAT</td>\n",
       "      <td>|||||||||||||</td>\n",
       "      <td>minCED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_1</td>\n",
       "      <td>LTBA01000001.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>290961</td>\n",
       "      <td>290972</td>\n",
       "      <td>2.549850</td>\n",
       "      <td>24.2811</td>\n",
       "      <td>30</td>\n",
       "      <td>0.400</td>\n",
       "      <td>+</td>\n",
       "      <td>TAACTTTTGTTA</td>\n",
       "      <td>TAACTTTTGTTA</td>\n",
       "      <td>||||||||||||</td>\n",
       "      <td>minCED</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               asm                    qseqid          sseqid  pident  length  \\\n",
       "0  GCA_001594005.1  GCA_001594005.1_m_1_c1_1  LTBA01000010.1   100.0      13   \n",
       "1  GCA_001594005.1  GCA_001594005.1_m_1_c1_1  LTBA01000010.1   100.0      12   \n",
       "2  GCA_001594005.1  GCA_001594005.1_m_1_c1_1  LTBA01000001.1   100.0      13   \n",
       "3  GCA_001594005.1  GCA_001594005.1_m_1_c1_1  LTBA01000001.1   100.0      13   \n",
       "4  GCA_001594005.1  GCA_001594005.1_m_1_c1_1  LTBA01000001.1   100.0      12   \n",
       "\n",
       "   mismatch  gap  qstart  qend  sstart    send    evalue  bitscore  qlen  \\\n",
       "0         0    0       8    20   10686   10674  0.645306   26.2635    30   \n",
       "1         0    0      10    21   52349   52338  2.549850   24.2811    30   \n",
       "2         0    0      10    22  267085  267097  0.645306   26.2635    30   \n",
       "3         0    0      12    24  344404  344392  0.645306   26.2635    30   \n",
       "4         0    0       1    12  290961  290972  2.549850   24.2811    30   \n",
       "\n",
       "    qcov strand           qseq           sseq            alm    tool  \n",
       "0  0.433      -  TGTTATGATAGGA  TGTTATGATAGGA  |||||||||||||  minCED  \n",
       "1  0.400      -   TTATGATAGGAG   TTATGATAGGAG   ||||||||||||  minCED  \n",
       "2  0.433      +  TTATGATAGGAGG  TTATGATAGGAGG  |||||||||||||  minCED  \n",
       "3  0.433      -  ATGATAGGAGGAT  ATGATAGGAGGAT  |||||||||||||  minCED  \n",
       "4  0.400      +   TAACTTTTGTTA   TAACTTTTGTTA   ||||||||||||  minCED  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Blast res extraction for consensus repeats (reordered)\n",
    "print(\">\", \"Blast res extraction for consensus repeats (reordered)\")\n",
    "if not os.path.exists(all_dir + \"/blastConReo.txt\"):\n",
    "    blcReop = pd.DataFrame()\n",
    "    for fl in sorted(os.listdir(blcReo_dir)):\n",
    "        asm = os.path.splitext(fl)[0]\n",
    "        #print(\">\", asm)\n",
    "        blcReop = pd.concat([blcReop, readBLAST(os.path.join(blcReo_dir, fl))], ignore_index=True)\n",
    "    blcReop[\"tool\"] = \"NA\"\n",
    "    blcReop.loc[blcReop[\"qseqid\"].str.contains(\"_p_\"), \"tool\"] = \"pilerCR\"\n",
    "    blcReop.loc[blcReop[\"qseqid\"].str.contains(\"_m_\"), \"tool\"] = \"minCED\"\n",
    "    blcReop.to_csv(all_dir + \"/blastConReo.txt\", header=True, sep=\"\\t\", quoting=csv.QUOTE_NONE, index=False)\n",
    "else:\n",
    "    print(\"{} exists, read it directly.\".format(all_dir + \"/blastConReo.txt\"))\n",
    "    blcReop = pd.read_csv(all_dir + \"/blastConReo.txt\", header=0, sep=\"\\t\", quoting=csv.QUOTE_NONE)\n",
    "print(blcReop.shape)\n",
    "blcReop.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "59bd46be-f739-4b1c-9ad6-a96b826114f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# stat\n",
    "# quantile by tool\n",
    "stat_qt = blcReop.groupby(\"tool\")[[\"pident\", \"qcov\"]].quantile([i / 100 for i in range(0, 101)]).reset_index()\n",
    "stat_qt = stat_qt.rename(columns={\"level_1\": \"prob\"})\n",
    "stat_qt95 = stat_qt.loc[stat_qt[\"prob\"] == 0.95, ]\n",
    "#stat_qt95.to_csv(all_dir + \"/blastConReo_stat_qt95.txt\", header=True, sep=\"\\t\", index=False)\n",
    "gp1 = ggplot(stat_qt, aes(x=\"pident\", y=\"prob\", group=\"tool\", color=\"tool\")) + geom_line() + geom_point(data=stat_qt.loc[stat_qt[\"prob\"] == 0.95, ]) + \\\n",
    "             xlab(\"Identity (%)\") + ylab(\"CDF\") + ggtitle(\"Background (reorder)\") + theme(aspect_ratio=1)\n",
    "gp2 = ggplot(stat_qt, aes(x=\"qcov\", y=\"prob\", group=\"tool\", color=\"tool\")) + geom_line() + geom_point(data=stat_qt.loc[stat_qt[\"prob\"] == 0.95, ]) + \\\n",
    "             xlab(\"Coverage in query\") + ylab(\"CDF\") + ggtitle(\"Background (reorder)\") + theme(aspect_ratio=1)\n",
    "# # quantile by tool and query length\n",
    "# stat_qt95_tl = blcReop.pivot_table(index=[\"tool\", \"qlen\"], values=[\"pident\", \"qcov\"], aggfunc=lambda x: x.quantile(0.95)).reset_index()\n",
    "# gp3 = ggplot(stat_qt95_tl) + geom_line(aes(x=\"qlen\", y=\"pident\", color=\"tool\")) + xlab(\"Query length\") + ylab(\"Identity\") + labs(color=\"Tool\") + \\\n",
    "#       theme(aspect_ratio=1)\n",
    "# gp4 = ggplot(stat_qt95_tl) + geom_line(aes(x=\"qlen\", y=\"qcov\", color=\"tool\")) + xlab(\"Query length\") + ylab(\"Coverage in query\") + labs(color=\"Tool\") + \\\n",
    "#       theme(aspect_ratio=1)\n",
    "# quantile by tool and query length (after average by query)\n",
    "stat_by_q = blcReop.pivot_table(index=[\"tool\", \"qseqid\", \"qlen\"], values=[\"pident\", \"qcov\"], aggfunc=lambda x: x.mean()).reset_index()\n",
    "stat_qt95_tla = stat_by_q.pivot_table(index=[\"tool\", \"qlen\"], values=[\"pident\", \"qcov\"], aggfunc=lambda x: x.quantile(0.95)).reset_index()\n",
    "gp3 = ggplot(stat_qt95_tla) + geom_line(aes(x=\"qlen\", y=\"pident\", color=\"tool\")) + scale_y_continuous(limits=(90, 100)) + \\\n",
    "      xlab(\"Query length\") + ylab(\"Identity\") + labs(color=\"Tool\") + theme(aspect_ratio=1)\n",
    "gp4 = ggplot(stat_qt95_tla) + geom_line(aes(x=\"qlen\", y=\"qcov\", color=\"tool\")) + \\\n",
    "      xlab(\"Query length\") + ylab(\"Coverage in query\") + labs(color=\"Tool\") + theme(aspect_ratio=1)\n",
    "# # comparison\n",
    "# stat_cmp = pd.merge(stat_qt95_tl, stat_qt95_tla, on=\"qlen\")\n",
    "# gp_cmp1 = ggplot(stat_cmp) + geom_point(aes(x=\"pident_x\", y=\"pident_y\"), alpha=0.6) + \\\n",
    "#     scale_x_continuous(limits=(90, 100)) + scale_y_continuous(limits=(90, 100)) + theme(aspect_ratio=1)\n",
    "# gp_cmp2 = ggplot(stat_cmp) + geom_point(aes(x=\"qcov_x\", y=\"qcov_y\"), alpha=0.6) + geom_abline(slope=1, intercept=0, color=\"grey\", linetype=\"dashed\") + \\\n",
    "#     scale_x_continuous(limits=(0, 1)) + scale_y_continuous(limits=(0, 1)) + theme(aspect_ratio=1)  # qt95 in stat_qt95_tla: small\n",
    "# quantile by query length\n",
    "stat_qt95_l = stat_by_q.pivot_table(index=[\"qlen\"], values=[\"pident\", \"qcov\"], aggfunc=lambda x: x.quantile(0.95)).reset_index()\n",
    "gp5 = ggplot(stat_qt95_l) + geom_line(aes(x=\"qlen\", y=\"pident\"), color=\"cornflowerblue\") + xlab(\"Query length\") + ylab(\"Identity\") + labs(color=\"Tool\") + \\\n",
    "      theme(aspect_ratio=1)\n",
    "gp6 = ggplot(stat_qt95_l) + geom_line(aes(x=\"qlen\", y=\"qcov\"), color=\"cornflowerblue\") + xlab(\"Query length\") + ylab(\"Coverage in query\") + labs(color=\"Tool\") + \\\n",
    "      theme(aspect_ratio=1)\n",
    "# constant mapping (pident ~ qlen)\n",
    "stat_qt95_l[\"pident_pred\"] = stat_qt95_l[\"pident\"].mean()\n",
    "# regression (qcov ~ qlen)\n",
    "def func(x, a, b, c):\n",
    "    return a ** (x + b) + c\n",
    "popt, pcov = curve_fit(func, stat_qt95_l[\"qlen\"], stat_qt95_l[\"qcov\"])\n",
    "stat_qt95_l[\"qcov_pred\"] = func(stat_qt95_l[\"qlen\"], *popt)\n",
    "stat_qt95_l.to_csv(all_dir + \"/blastConReo_stat_qt95.txt\", header=True, sep=\"\\t\", index=False)\n",
    "gp7 = ggplot(stat_qt95_l) + geom_line(aes(x=\"qlen\", y=\"qcov\"), color=\"cornflowerblue\") + geom_line(aes(x=\"qlen\", y=\"qcov_pred\"), color=\"black\", size=1) + \\\n",
    "      xlab(\"Query length\") + ylab(\"Coverage in query\") + labs(color=\"Tool\") + \\\n",
    "      theme(aspect_ratio=1)\n",
    "gps = [gp1, gp2, gp3, gp4, gp5, gp6, gp7]\n",
    "save_as_pdf_pages(gps, all_dir + \"/blastConReo_stat.pdf\", verbose=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1e539cb5-aaa9-410a-b5cd-83dea896ee29",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> Blast for consensus repeats\n",
      "12 / 12 runs were skipped.\n"
     ]
    }
   ],
   "source": [
    "# Blast for consensus repeats\n",
    "print(\">\", \"Blast for consensus repeats\")\n",
    "blc_dir = output_dir + \"/blastCon\"\n",
    "if not os.path.exists(blc_dir):\n",
    "    os.makedirs(blc_dir)\n",
    "skipped = 0\n",
    "fls = sorted(os.listdir(con_dir))\n",
    "for fl in fls:\n",
    "    asm = os.path.splitext(fl)[0]\n",
    "    #print(\">\", asm)\n",
    "    asm_file = os.path.join(asm_o_dir, fl)\n",
    "    xml_file = os.path.join(blc_dir, asm + \".xml\")\n",
    "    if not os.path.exists(xml_file):\n",
    "        os.system(\"blastn -query {} -db {} -task blastn-short -dust no -outfmt 5 -num_threads 12 -out {}\".format(os.path.join(con_dir, fl), asm_file, xml_file))  # old para\n",
    "    else:\n",
    "        skipped += 1\n",
    "        #print(\"{} exists, skipped.\".format(xml_file))\n",
    "if skipped > 0:\n",
    "    print(\"{} / {} runs were skipped.\".format(skipped, len(fls)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8ab86eae-45d8-4fea-bf58-2f5d8b6f227e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> Blast res extraction for consensus repeats\n",
      "(43940, 20)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>asm</th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>pident</th>\n",
       "      <th>length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gap</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>qlen</th>\n",
       "      <th>qcov</th>\n",
       "      <th>strand</th>\n",
       "      <th>qseq</th>\n",
       "      <th>sseq</th>\n",
       "      <th>alm</th>\n",
       "      <th>tool</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>GCA_001594005.1_m_1_c1</td>\n",
       "      <td>LTBA01000005.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>79689</td>\n",
       "      <td>79718</td>\n",
       "      <td>4.624340e-11</td>\n",
       "      <td>59.9635</td>\n",
       "      <td>30</td>\n",
       "      <td>1.000</td>\n",
       "      <td>+</td>\n",
       "      <td>ATAAATGGTTCAGGGGATATTAATGGTGAT</td>\n",
       "      <td>ATAAATGGTTCAGGGGATATTAATGGTGAT</td>\n",
       "      <td>||||||||||||||||||||||||||||||</td>\n",
       "      <td>minCED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>GCA_001594005.1_m_1_c1</td>\n",
       "      <td>LTBA01000005.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>27</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>27</td>\n",
       "      <td>79740</td>\n",
       "      <td>79766</td>\n",
       "      <td>2.852950e-09</td>\n",
       "      <td>54.0164</td>\n",
       "      <td>30</td>\n",
       "      <td>0.900</td>\n",
       "      <td>+</td>\n",
       "      <td>ATAAATGGTTCAGGGGATATTAATGGT</td>\n",
       "      <td>ATAAATGGTTCAGGGGATATTAATGGT</td>\n",
       "      <td>|||||||||||||||||||||||||||</td>\n",
       "      <td>minCED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>GCA_001594005.1_m_1_c1</td>\n",
       "      <td>LTBA01000005.1</td>\n",
       "      <td>96.154</td>\n",
       "      <td>26</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>26</td>\n",
       "      <td>79791</td>\n",
       "      <td>79816</td>\n",
       "      <td>2.748110e-06</td>\n",
       "      <td>44.1047</td>\n",
       "      <td>30</td>\n",
       "      <td>0.867</td>\n",
       "      <td>+</td>\n",
       "      <td>ATAAATGGTTCAGGGGATATTAATGG</td>\n",
       "      <td>ATAAATGGTTCAGGAGATATTAATGG</td>\n",
       "      <td>|||||||||||||| |||||||||||</td>\n",
       "      <td>minCED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>GCA_001594005.1_m_1_c1</td>\n",
       "      <td>LTBA01000005.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>79638</td>\n",
       "      <td>79657</td>\n",
       "      <td>4.290720e-05</td>\n",
       "      <td>40.1400</td>\n",
       "      <td>30</td>\n",
       "      <td>0.667</td>\n",
       "      <td>+</td>\n",
       "      <td>ATAAATGGTTCAGGGGATAT</td>\n",
       "      <td>ATAAATGGTTCAGGGGATAT</td>\n",
       "      <td>||||||||||||||||||||</td>\n",
       "      <td>minCED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>GCA_001594005.1_m_1_c1</td>\n",
       "      <td>LTBA01000005.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>14</td>\n",
       "      <td>38212</td>\n",
       "      <td>38223</td>\n",
       "      <td>2.549850e+00</td>\n",
       "      <td>24.2811</td>\n",
       "      <td>30</td>\n",
       "      <td>0.400</td>\n",
       "      <td>+</td>\n",
       "      <td>AAATGGTTCAGG</td>\n",
       "      <td>AAATGGTTCAGG</td>\n",
       "      <td>||||||||||||</td>\n",
       "      <td>minCED</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               asm                  qseqid          sseqid   pident  length  \\\n",
       "0  GCA_001594005.1  GCA_001594005.1_m_1_c1  LTBA01000005.1  100.000      30   \n",
       "1  GCA_001594005.1  GCA_001594005.1_m_1_c1  LTBA01000005.1  100.000      27   \n",
       "2  GCA_001594005.1  GCA_001594005.1_m_1_c1  LTBA01000005.1   96.154      26   \n",
       "3  GCA_001594005.1  GCA_001594005.1_m_1_c1  LTBA01000005.1  100.000      20   \n",
       "4  GCA_001594005.1  GCA_001594005.1_m_1_c1  LTBA01000005.1  100.000      12   \n",
       "\n",
       "   mismatch  gap  qstart  qend  sstart   send        evalue  bitscore  qlen  \\\n",
       "0         0    0       1    30   79689  79718  4.624340e-11   59.9635    30   \n",
       "1         0    0       1    27   79740  79766  2.852950e-09   54.0164    30   \n",
       "2         1    0       1    26   79791  79816  2.748110e-06   44.1047    30   \n",
       "3         0    0       1    20   79638  79657  4.290720e-05   40.1400    30   \n",
       "4         0    0       3    14   38212  38223  2.549850e+00   24.2811    30   \n",
       "\n",
       "    qcov strand                            qseq  \\\n",
       "0  1.000      +  ATAAATGGTTCAGGGGATATTAATGGTGAT   \n",
       "1  0.900      +     ATAAATGGTTCAGGGGATATTAATGGT   \n",
       "2  0.867      +      ATAAATGGTTCAGGGGATATTAATGG   \n",
       "3  0.667      +            ATAAATGGTTCAGGGGATAT   \n",
       "4  0.400      +                    AAATGGTTCAGG   \n",
       "\n",
       "                             sseq                             alm    tool  \n",
       "0  ATAAATGGTTCAGGGGATATTAATGGTGAT  ||||||||||||||||||||||||||||||  minCED  \n",
       "1     ATAAATGGTTCAGGGGATATTAATGGT     |||||||||||||||||||||||||||  minCED  \n",
       "2      ATAAATGGTTCAGGAGATATTAATGG      |||||||||||||| |||||||||||  minCED  \n",
       "3            ATAAATGGTTCAGGGGATAT            ||||||||||||||||||||  minCED  \n",
       "4                    AAATGGTTCAGG                    ||||||||||||  minCED  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Blast res extraction for consensus repeats\n",
    "print(\">\", \"Blast res extraction for consensus repeats\")\n",
    "blcp = pd.DataFrame()\n",
    "for fl in sorted(os.listdir(blc_dir)):\n",
    "    asm = os.path.splitext(fl)[0]\n",
    "    #print(\">\", asm)\n",
    "    blcp = pd.concat([blcp, readBLAST(os.path.join(blc_dir, fl))], ignore_index=True)  \n",
    "blcp[\"tool\"] = \"NA\"\n",
    "blcp.loc[blcp[\"qseqid\"].str.contains(\"_p_\"), \"tool\"] = \"pilerCR\"\n",
    "blcp.loc[blcp[\"qseqid\"].str.contains(\"_m_\"), \"tool\"] = \"minCED\"\n",
    "blcp.to_csv(all_dir + \"/blastCon.txt\", header=True, sep=\"\\t\", quoting=csv.QUOTE_NONE, index=False)\n",
    "print(blcp.shape)\n",
    "blcp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "28c032fc-48a3-4066-af92-c5b3f9aa08bf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100.00% (237 / 237) of consensus repeats have hits.\n"
     ]
    }
   ],
   "source": [
    "# stat (query length ~ hit number)\n",
    "stat = blcp.pivot_table(index=[\"asm\", \"qseqid\", \"qlen\"], values=\"sseqid\", aggfunc=len).reset_index().rename(columns={\"sseqid\": \"hit_num\"})\n",
    "stat = pd.merge(conRes, stat[[\"qseqid\", \"qlen\", \"hit_num\"]], left_on=\"id\", right_on=\"qseqid\", how=\"left\")\n",
    "stat[\"length\"] = [float(i) for i in stat[\"length\"]]\n",
    "stat.loc[stat[\"hit_num\"].isna(), \"hit_num\"] = 0\n",
    "stat[\"hit_log10\"] = [math.log10(i + 1) for i in stat[\"hit_num\"]]\n",
    "stat.to_csv(all_dir + \"/blastCon_stat.txt\", header=True, sep=\"\\t\", quoting=csv.QUOTE_NONE, index=False)\n",
    "#print(stat.shape)\n",
    "#stat.head()\n",
    "# plot\n",
    "gp = ggplot(stat) + geom_vline(xintercept=17.5, color=\"grey\", linetype=\"dashed\") + geom_point(aes(x=\"length\", y=\"hit_log10\", color=\"asm\"), alpha=0.6) + \\\n",
    "     facet_wrap(\"~asm\", ncol=2) + scale_color_discrete(guide=False) + scale_x_continuous() + scale_y_continuous(limits=[0, stat[\"hit_log10\"].max()]) +\\\n",
    "     xlab(\"Consensus repeat length\") + ylab(\"Blast hit number (log10 + 1)\")\n",
    "gp.save((all_dir + \"/blastCon_stat.pdf\"), width=6, height=6, verbose=False)\n",
    "# whether con repeats have hits\n",
    "con_wht_num = len(stat[stat[\"hit_num\"] > 0][\"id\"].unique())\n",
    "con_tot_num = stat.shape[0]\n",
    "print(\"{:.2%} ({} / {}) of consensus repeats have hits.\".format(round(con_wht_num / con_tot_num, 3), con_wht_num, con_tot_num))\n",
    "if con_wht_num != con_tot_num:\n",
    "    print(\"Missed con repeats:\")\n",
    "    print(stat[stat[\"hit_num\"] == 0].pivot_table(index=\"length\", values=\"id\", aggfunc=len).reset_index().rename(columns={\"id\": \"con_num\"}))\n",
    "    stat[stat[\"hit_num\"] == 0].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2f489416-198d-4f5f-b8b7-504aeeaaf402",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> Blast res filtering\n",
      ">> Filtering (pass the cutoff of pident and qcov)...\n",
      "15029 / 43940 hits passed.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>asm</th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>pident</th>\n",
       "      <th>length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gap</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>...</th>\n",
       "      <th>qcov</th>\n",
       "      <th>strand</th>\n",
       "      <th>qseq</th>\n",
       "      <th>sseq</th>\n",
       "      <th>alm</th>\n",
       "      <th>tool</th>\n",
       "      <th>pident_qt95</th>\n",
       "      <th>qcov_qt95</th>\n",
       "      <th>pident_pred</th>\n",
       "      <th>qcov_pred</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>GCA_001594005.1_m_1_c1</td>\n",
       "      <td>LTBA01000005.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>79689</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000</td>\n",
       "      <td>+</td>\n",
       "      <td>ATAAATGGTTCAGGGGATATTAATGGTGAT</td>\n",
       "      <td>ATAAATGGTTCAGGGGATATTAATGGTGAT</td>\n",
       "      <td>||||||||||||||||||||||||||||||</td>\n",
       "      <td>minCED</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.449455</td>\n",
       "      <td>99.973229</td>\n",
       "      <td>0.447913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>GCA_001594005.1_m_1_c1</td>\n",
       "      <td>LTBA01000005.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>27</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>27</td>\n",
       "      <td>79740</td>\n",
       "      <td>...</td>\n",
       "      <td>0.900</td>\n",
       "      <td>+</td>\n",
       "      <td>ATAAATGGTTCAGGGGATATTAATGGT</td>\n",
       "      <td>ATAAATGGTTCAGGGGATATTAATGGT</td>\n",
       "      <td>|||||||||||||||||||||||||||</td>\n",
       "      <td>minCED</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.449455</td>\n",
       "      <td>99.973229</td>\n",
       "      <td>0.447913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>GCA_001594005.1_m_1_c1</td>\n",
       "      <td>LTBA01000005.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>79638</td>\n",
       "      <td>...</td>\n",
       "      <td>0.667</td>\n",
       "      <td>+</td>\n",
       "      <td>ATAAATGGTTCAGGGGATAT</td>\n",
       "      <td>ATAAATGGTTCAGGGGATAT</td>\n",
       "      <td>||||||||||||||||||||</td>\n",
       "      <td>minCED</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.449455</td>\n",
       "      <td>99.973229</td>\n",
       "      <td>0.447913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>GCA_001594005.1_m_1_c1</td>\n",
       "      <td>LTBA01000045.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>24</td>\n",
       "      <td>15665</td>\n",
       "      <td>...</td>\n",
       "      <td>0.500</td>\n",
       "      <td>-</td>\n",
       "      <td>TCAGGGGATATTAAT</td>\n",
       "      <td>TCAGGGGATATTAAT</td>\n",
       "      <td>|||||||||||||||</td>\n",
       "      <td>minCED</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.449455</td>\n",
       "      <td>99.973229</td>\n",
       "      <td>0.447913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>GCA_001594005.1_m_1_c1</td>\n",
       "      <td>LTBA01000021.1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>29</td>\n",
       "      <td>28999</td>\n",
       "      <td>...</td>\n",
       "      <td>0.467</td>\n",
       "      <td>+</td>\n",
       "      <td>GATATTAATGGTGA</td>\n",
       "      <td>GATATTAATGGTGA</td>\n",
       "      <td>||||||||||||||</td>\n",
       "      <td>minCED</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.449455</td>\n",
       "      <td>99.973229</td>\n",
       "      <td>0.447913</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               asm                  qseqid          sseqid  pident  length  \\\n",
       "0  GCA_001594005.1  GCA_001594005.1_m_1_c1  LTBA01000005.1   100.0      30   \n",
       "1  GCA_001594005.1  GCA_001594005.1_m_1_c1  LTBA01000005.1   100.0      27   \n",
       "3  GCA_001594005.1  GCA_001594005.1_m_1_c1  LTBA01000005.1   100.0      20   \n",
       "6  GCA_001594005.1  GCA_001594005.1_m_1_c1  LTBA01000045.1   100.0      15   \n",
       "7  GCA_001594005.1  GCA_001594005.1_m_1_c1  LTBA01000021.1   100.0      14   \n",
       "\n",
       "   mismatch  gap  qstart  qend  sstart  ...   qcov  strand  \\\n",
       "0         0    0       1    30   79689  ...  1.000       +   \n",
       "1         0    0       1    27   79740  ...  0.900       +   \n",
       "3         0    0       1    20   79638  ...  0.667       +   \n",
       "6         0    0      10    24   15665  ...  0.500       -   \n",
       "7         0    0      16    29   28999  ...  0.467       +   \n",
       "\n",
       "                             qseq                            sseq  \\\n",
       "0  ATAAATGGTTCAGGGGATATTAATGGTGAT  ATAAATGGTTCAGGGGATATTAATGGTGAT   \n",
       "1     ATAAATGGTTCAGGGGATATTAATGGT     ATAAATGGTTCAGGGGATATTAATGGT   \n",
       "3            ATAAATGGTTCAGGGGATAT            ATAAATGGTTCAGGGGATAT   \n",
       "6                 TCAGGGGATATTAAT                 TCAGGGGATATTAAT   \n",
       "7                  GATATTAATGGTGA                  GATATTAATGGTGA   \n",
       "\n",
       "                              alm    tool pident_qt95 qcov_qt95 pident_pred  \\\n",
       "0  ||||||||||||||||||||||||||||||  minCED       100.0  0.449455   99.973229   \n",
       "1     |||||||||||||||||||||||||||  minCED       100.0  0.449455   99.973229   \n",
       "3            ||||||||||||||||||||  minCED       100.0  0.449455   99.973229   \n",
       "6                 |||||||||||||||  minCED       100.0  0.449455   99.973229   \n",
       "7                  ||||||||||||||  minCED       100.0  0.449455   99.973229   \n",
       "\n",
       "  qcov_pred  \n",
       "0  0.447913  \n",
       "1  0.447913  \n",
       "3  0.447913  \n",
       "6  0.447913  \n",
       "7  0.447913  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Blast res filtering\n",
    "print(\">\", \"Blast res filtering\")\n",
    "#stat_qt95_l = pd.read_csv(all_dir + \"/blastConReo_stat_qt95.txt\", header=0, sep=\"\\t\")\n",
    "blcp_ct = pd.merge(blcp, stat_qt95_l, on=\"qlen\", suffixes=[\"\", \"_qt95\"])\n",
    "# blcp_ct = pd.merge(blcp_ct, conRes[[\"id\", \"cid\", \"seq\"]], left_on=\"qseqid\", right_on=\"id\").drop(columns=\"id\")\n",
    "# blcp_ct = pd.merge(blcp_ct, criPos, left_on=\"cid\", right_on=\"cid\")\n",
    "# cads = []\n",
    "# for _, row in blcp_ct.iterrows():\n",
    "#     if row[\"contig\"] == row[\"sseqid\"]:\n",
    "#         ant_l, ant_r = row[\"sstart\"] - 1, row[\"send\"]\n",
    "#         if row[\"end\"] <= ant_l:\n",
    "#             cad = ant_l - row[\"end\"]\n",
    "#         elif row[\"start\"] >= ant_r:\n",
    "#             cad = - (row[\"start\"] - ant_r)\n",
    "#         else:\n",
    "#             cad = 0\n",
    "#     else:\n",
    "#         cad = int(1e9)\n",
    "#     cads.append(cad)\n",
    "# blcp_ct[\"cadist\"] = cads\n",
    "# cutoff: pident and qcov\n",
    "print(\">> Filtering (pass the cutoff of pident and qcov)...\")\n",
    "blcp_ct_ftd = blcp_ct.loc[(blcp_ct[\"pident\"] >= blcp_ct[\"pident_pred\"]) & (blcp_ct[\"qcov\"] >= blcp_ct[\"qcov_pred\"])]\n",
    "print(\"{} / {} hits passed.\".format(blcp_ct_ftd.shape[0], blcp_ct.shape[0]))\n",
    "# cutoff: in the same contig\n",
    "# print(\">> Filtering (in the same contig)...\")\n",
    "# blcp_ct_ftd = blcp_ct_ftd[blcp_ct_ftd[\"contig\"] == blcp_ct_ftd[\"sseqid\"]]\n",
    "# print(\"{} / {} hits passed the filitering.\".format(blcp_ct_ftd.shape[0], blcp_ct.shape[0]))\n",
    "# cutoff: away from CRISPR arrays (corresponding CRISPR array)\n",
    "# print(\">> Filtering (away from CRISPR arrays)...\")\n",
    "# blcp_ct_ftd = blcp_ct_ftd[blcp_ct_ftd[\"cadist\"].abs() > 0]\n",
    "# print(\"{} / {} hits passed.\".format(blcp_ct_ftd.shape[0], blcp_ct.shape[0]))\n",
    "# write\n",
    "blcp_ct_ftd.to_csv(all_dir + \"/blastCon_ftd.txt\", header=True, sep=\"\\t\", quoting=csv.QUOTE_NONE, index=False)\n",
    "blcp_ct_ftd.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "790a09d6-b9ce-40ad-a90e-a351a48cb60a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(237, 11)\n",
      "99.20% (235 / 237) of con repeats have hits.\n",
      "Missed con repeats:\n",
      "   length  con_num\n",
      "0    51.0        1\n",
      "1    53.0        1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>asm</th>\n",
       "      <th>tool</th>\n",
       "      <th>id</th>\n",
       "      <th>cid</th>\n",
       "      <th>seq</th>\n",
       "      <th>length</th>\n",
       "      <th>qseqid</th>\n",
       "      <th>qlen</th>\n",
       "      <th>hit_num</th>\n",
       "      <th>hit_log10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>RGDI01000039.1</td>\n",
       "      <td>GCA_009918665.1</td>\n",
       "      <td>pilerCR</td>\n",
       "      <td>GCA_009918665.1_p_3_c1</td>\n",
       "      <td>GCA_009918665.1_p_3</td>\n",
       "      <td>TATGACTATGGCTTGAACTAGCATAACTGCCAGATGGTTGTTTACC...</td>\n",
       "      <td>53.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>RGDD01000047.1</td>\n",
       "      <td>GCA_009918765.1</td>\n",
       "      <td>pilerCR</td>\n",
       "      <td>GCA_009918765.1_p_2_c1</td>\n",
       "      <td>GCA_009918765.1_p_2</td>\n",
       "      <td>TCTAGATGGTAAACAACCATCTGGCAGTTATGCTAATTCAAGCCAT...</td>\n",
       "      <td>51.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            contig              asm     tool                      id  \\\n",
       "31  RGDI01000039.1  GCA_009918665.1  pilerCR  GCA_009918665.1_p_3_c1   \n",
       "48  RGDD01000047.1  GCA_009918765.1  pilerCR  GCA_009918765.1_p_2_c1   \n",
       "\n",
       "                    cid                                                seq  \\\n",
       "31  GCA_009918665.1_p_3  TATGACTATGGCTTGAACTAGCATAACTGCCAGATGGTTGTTTACC...   \n",
       "48  GCA_009918765.1_p_2  TCTAGATGGTAAACAACCATCTGGCAGTTATGCTAATTCAAGCCAT...   \n",
       "\n",
       "    length qseqid  qlen  hit_num  hit_log10  \n",
       "31    53.0    NaN   NaN      0.0        0.0  \n",
       "48    51.0    NaN   NaN      0.0        0.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# stat (query length ~ hit number)\n",
    "stat_ftd = blcp_ct_ftd.pivot_table(index=[\"asm\", \"qseqid\", \"qlen\"], values=\"sseqid\", aggfunc=len).reset_index().rename(columns={\"sseqid\": \"hit_num\"})\n",
    "stat_ftd = pd.merge(conRes, stat_ftd[[\"qseqid\", \"qlen\", \"hit_num\"]], left_on=\"id\", right_on=\"qseqid\", how=\"left\")\n",
    "stat_ftd[\"length\"] = [float(i) for i in stat_ftd[\"length\"]]\n",
    "stat_ftd.loc[stat_ftd[\"hit_num\"].isna(), \"hit_num\"] = 0\n",
    "stat_ftd[\"hit_log10\"] = [math.log10(i + 1) for i in stat_ftd[\"hit_num\"]]\n",
    "stat_ftd.to_csv(all_dir + \"/blastCon_stat_ftd.txt\", header=True, sep=\"\\t\", quoting=csv.QUOTE_NONE, index=False)\n",
    "print(stat_ftd.shape)\n",
    "#stat_ftd.head()\n",
    "# plot\n",
    "gp = ggplot(stat_ftd) + geom_vline(xintercept=17.5, color=\"grey\", linetype=\"dashed\") + geom_point(aes(x=\"length\", y=\"hit_log10\", color=\"asm\"), alpha=0.6) + \\\n",
    "     facet_wrap(\"~asm\", ncol=2) + scale_color_discrete(guide=False) + scale_x_continuous() + scale_y_continuous(limits=[0, stat_ftd[\"hit_log10\"].max()]) +\\\n",
    "     xlab(\"Consensus repeat length\") + ylab(\"Blast hit number (log10 + 1)\")\n",
    "gp.save((all_dir + \"/blastCon_stat_ftd.pdf\"), width=6, height=6, verbose=False)\n",
    "# whether con repeats have hits\n",
    "con_wht_num = len(stat_ftd[stat_ftd[\"hit_num\"] > 0][\"id\"].unique())\n",
    "con_tot_num = stat_ftd.shape[0]\n",
    "print(\"{:.2%} ({} / {}) of con repeats have hits.\".format(round(con_wht_num / con_tot_num, 3), con_wht_num, con_tot_num))\n",
    "if con_wht_num != con_tot_num:\n",
    "    print(\"Missed con repeats:\")\n",
    "    print(stat_ftd[stat_ftd[\"hit_num\"] == 0].pivot_table(index=\"length\", values=\"id\", aggfunc=len).reset_index().rename(columns={\"id\": \"con_num\"}))\n",
    "    display(stat_ftd[stat_ftd[\"hit_num\"] == 0].head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c432453d-9ec1-4938-98df-c5ea046cccb4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> Blast res -> Anti-repeat\n",
      "(15029, 12)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>id</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>asm</th>\n",
       "      <th>tool</th>\n",
       "      <th>conid</th>\n",
       "      <th>component</th>\n",
       "      <th>seq</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LTBA01000005.1</td>\n",
       "      <td>79688</td>\n",
       "      <td>79718</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a1</td>\n",
       "      <td>900</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>minCED</td>\n",
       "      <td>GCA_001594005.1_m_1_c1</td>\n",
       "      <td>antiRep</td>\n",
       "      <td>ATAAATGGTTCAGGGGATATTAATGGTGAT</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LTBA01000005.1</td>\n",
       "      <td>79739</td>\n",
       "      <td>79766</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a2</td>\n",
       "      <td>900</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>minCED</td>\n",
       "      <td>GCA_001594005.1_m_1_c1</td>\n",
       "      <td>antiRep</td>\n",
       "      <td>ATAAATGGTTCAGGGGATATTAATGGT</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LTBA01000005.1</td>\n",
       "      <td>79637</td>\n",
       "      <td>79657</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a3</td>\n",
       "      <td>900</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>minCED</td>\n",
       "      <td>GCA_001594005.1_m_1_c1</td>\n",
       "      <td>antiRep</td>\n",
       "      <td>ATAAATGGTTCAGGGGATAT</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LTBA01000045.1</td>\n",
       "      <td>15664</td>\n",
       "      <td>15679</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a4</td>\n",
       "      <td>900</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>minCED</td>\n",
       "      <td>GCA_001594005.1_m_1_c1</td>\n",
       "      <td>antiRep</td>\n",
       "      <td>TCAGGGGATATTAAT</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LTBA01000021.1</td>\n",
       "      <td>28998</td>\n",
       "      <td>29012</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a5</td>\n",
       "      <td>900</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>minCED</td>\n",
       "      <td>GCA_001594005.1_m_1_c1</td>\n",
       "      <td>antiRep</td>\n",
       "      <td>GATATTAATGGTGA</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           contig  start    end                         id  score strand  \\\n",
       "0  LTBA01000005.1  79688  79718  GCA_001594005.1_m_1_c1_a1    900      +   \n",
       "1  LTBA01000005.1  79739  79766  GCA_001594005.1_m_1_c1_a2    900      +   \n",
       "2  LTBA01000005.1  79637  79657  GCA_001594005.1_m_1_c1_a3    900      +   \n",
       "3  LTBA01000045.1  15664  15679  GCA_001594005.1_m_1_c1_a4    900      -   \n",
       "4  LTBA01000021.1  28998  29012  GCA_001594005.1_m_1_c1_a5    900      +   \n",
       "\n",
       "               asm    tool                   conid component  \\\n",
       "0  GCA_001594005.1  minCED  GCA_001594005.1_m_1_c1   antiRep   \n",
       "1  GCA_001594005.1  minCED  GCA_001594005.1_m_1_c1   antiRep   \n",
       "2  GCA_001594005.1  minCED  GCA_001594005.1_m_1_c1   antiRep   \n",
       "3  GCA_001594005.1  minCED  GCA_001594005.1_m_1_c1   antiRep   \n",
       "4  GCA_001594005.1  minCED  GCA_001594005.1_m_1_c1   antiRep   \n",
       "\n",
       "                              seq  length  \n",
       "0  ATAAATGGTTCAGGGGATATTAATGGTGAT      30  \n",
       "1     ATAAATGGTTCAGGGGATATTAATGGT      27  \n",
       "2            ATAAATGGTTCAGGGGATAT      20  \n",
       "3                 TCAGGGGATATTAAT      15  \n",
       "4                  GATATTAATGGTGA      14  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Blast res -> Anti-repeat\n",
    "print(\">\", \"Blast res -> Anti-repeat\")\n",
    "antRes = blcp_ct_ftd[[\"sseqid\", \"sstart\", \"send\", \"strand\", \"asm\", \"tool\", \"qseqid\", \"sseq\"]].reset_index(drop=True)\n",
    "antRes[\"sstart\"] = antRes[\"sstart\"] - 1\n",
    "antRes.insert(3, \"id\", \"NA\")\n",
    "antRes.insert(4, \"score\", 900)\n",
    "antRes.insert(9, \"component\", \"antiRep\")\n",
    "antRes.insert(11, \"length\", antRes[\"sseq\"].str.len())\n",
    "antRes = antRes.rename(columns={\"sseqid\": \"contig\", \"sstart\": \"start\", \"send\": \"end\", \"qseqid\": \"conid\", \"sseq\": \"seq\"})\n",
    "dfs = []\n",
    "for conid, df in antRes.groupby(\"conid\"):\n",
    "    df[\"id\"] = [conid + \"_a\" + str(i) for i in range(1, df.shape[0] + 1)]\n",
    "    dfs.append(df)\n",
    "antRes = pd.concat(dfs, ignore_index=True)\n",
    "antRes.to_csv(all_dir + \"/antiRep.txt\", header=True, sep=\"\\t\", quoting=csv.QUOTE_NONE, index=False)\n",
    "print(antRes.shape)\n",
    "antRes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e01102d8-4b9d-4f03-b747-9f5c315e21e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> CRISPR-close regions calculation...\n",
      "Expand CRISPRs regions by +/- 100 bp.\n"
     ]
    }
   ],
   "source": [
    "# CRISPR-close regions calc\n",
    "print(\">\", \"CRISPR-close regions calculation...\")\n",
    "slop_len = 100\n",
    "print(\"Expand CRISPRs regions by +/- {} bp.\".format(slop_len))\n",
    "criSlp = slopReg(criMet, \"lr\", slop_len, genomeDir=asm_o_dir)\n",
    "criSlp.to_csv(all_dir + \"/criSlp.txt\", header=True, sep=\"\\t\", quoting=csv.QUOTE_NONE, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "da40845a-c672-4b6f-9914-e394c8582ea0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> Anti-repeat filitering...\n",
      "(11095, 12)\n"
     ]
    }
   ],
   "source": [
    "# Anti-repeat filitering\n",
    "print(\">\", \"Anti-repeat filitering...\")\n",
    "os.system(\"/bin/bash -c \\\"bedtools subtract -A -a <(tail -n +2 {}) -b <(tail -n +2 {}) > {}\\\"\".format(all_dir + \"/antiRep.txt\", all_dir + \"/criSlp.txt\", all_dir + \"/antiRep_ftd.txt\"))\n",
    "antRes_ftd = pd.read_csv(all_dir + \"/antiRep_ftd.txt\", header=None, sep=\"\\t\", quoting=csv.QUOTE_NONE)\n",
    "antRes_ftd.columns = antRes.columns\n",
    "print(antRes_ftd.shape)\n",
    "#antRes_ftd.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "1d1e2d9f-cf57-4c66-ac3c-4f6ef1e24954",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> Anti-repeat duplicates removing...\n",
      "(10046, 12)\n"
     ]
    }
   ],
   "source": [
    "# Anti-repeat duplicates removing\n",
    "print(\">\", \"Anti-repeat duplicates removing...\")\n",
    "#antRes_ftd[\"occ_num\"] = antRes_ftd.groupby([\"asm\", \"contig\", \"start\", \"end\"])[\"id\"].transform(\"size\")\n",
    "#antRes_ftd[antRes_ftd[\"occ_num\"] > 1].shape\n",
    "antRes_ftd = antRes_ftd.drop_duplicates([\"asm\", \"contig\", \"start\", \"end\"])\n",
    "print(antRes_ftd.shape)\n",
    "#antRes_ftd.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "030feb97-95b6-41d6-abcf-5f22f50d74f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> Anti-repeat flanking regions calculation\n",
      "(20092, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>id</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>asm</th>\n",
       "      <th>tool</th>\n",
       "      <th>conid</th>\n",
       "      <th>component</th>\n",
       "      <th>seq</th>\n",
       "      <th>length</th>\n",
       "      <th>length_ctg</th>\n",
       "      <th>oid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LTBA01000045.1</td>\n",
       "      <td>15314</td>\n",
       "      <td>15664</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a4_l</td>\n",
       "      <td>900</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>minCED</td>\n",
       "      <td>GCA_001594005.1_m_1_c1</td>\n",
       "      <td>antiRepFlank</td>\n",
       "      <td>ATAACTCTTCCACATACTTCAGAATTTTATTTAGATGCTGAAGTTA...</td>\n",
       "      <td>350</td>\n",
       "      <td>17162</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LTBA01000045.1</td>\n",
       "      <td>15679</td>\n",
       "      <td>16029</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a4_r</td>\n",
       "      <td>900</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>minCED</td>\n",
       "      <td>GCA_001594005.1_m_1_c1</td>\n",
       "      <td>antiRepFlank</td>\n",
       "      <td>TGTACTTTTAATATTTATATCATTATTGAACGTTTCGTAATTCAAT...</td>\n",
       "      <td>350</td>\n",
       "      <td>17162</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LTBA01000021.1</td>\n",
       "      <td>28648</td>\n",
       "      <td>28998</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a5_l</td>\n",
       "      <td>900</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>minCED</td>\n",
       "      <td>GCA_001594005.1_m_1_c1</td>\n",
       "      <td>antiRepFlank</td>\n",
       "      <td>ACACACTTCATAAGGAGTATGTGAATAATCCCATATCTCTTGAAAC...</td>\n",
       "      <td>350</td>\n",
       "      <td>45775</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LTBA01000021.1</td>\n",
       "      <td>29012</td>\n",
       "      <td>29362</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a5_r</td>\n",
       "      <td>900</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>minCED</td>\n",
       "      <td>GCA_001594005.1_m_1_c1</td>\n",
       "      <td>antiRepFlank</td>\n",
       "      <td>GTGGGCGTTATTAAAAAGATAAAAAAAGAGAAAAAATATAAGGCAT...</td>\n",
       "      <td>350</td>\n",
       "      <td>45775</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LTBA01000010.1</td>\n",
       "      <td>48216</td>\n",
       "      <td>48566</td>\n",
       "      <td>GCA_001594005.1_m_2_c1_a3_l</td>\n",
       "      <td>900</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>minCED</td>\n",
       "      <td>GCA_001594005.1_m_2_c1</td>\n",
       "      <td>antiRepFlank</td>\n",
       "      <td>TCTTTTAGAAATATTAATATTCATAACCTTACTCAATTTATCTAAT...</td>\n",
       "      <td>350</td>\n",
       "      <td>79404</td>\n",
       "      <td>GCA_001594005.1_m_2_c1_a3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           contig  start    end                           id  score strand  \\\n",
       "0  LTBA01000045.1  15314  15664  GCA_001594005.1_m_1_c1_a4_l    900      -   \n",
       "0  LTBA01000045.1  15679  16029  GCA_001594005.1_m_1_c1_a4_r    900      +   \n",
       "1  LTBA01000021.1  28648  28998  GCA_001594005.1_m_1_c1_a5_l    900      -   \n",
       "1  LTBA01000021.1  29012  29362  GCA_001594005.1_m_1_c1_a5_r    900      +   \n",
       "2  LTBA01000010.1  48216  48566  GCA_001594005.1_m_2_c1_a3_l    900      -   \n",
       "\n",
       "               asm    tool                   conid     component  \\\n",
       "0  GCA_001594005.1  minCED  GCA_001594005.1_m_1_c1  antiRepFlank   \n",
       "0  GCA_001594005.1  minCED  GCA_001594005.1_m_1_c1  antiRepFlank   \n",
       "1  GCA_001594005.1  minCED  GCA_001594005.1_m_1_c1  antiRepFlank   \n",
       "1  GCA_001594005.1  minCED  GCA_001594005.1_m_1_c1  antiRepFlank   \n",
       "2  GCA_001594005.1  minCED  GCA_001594005.1_m_2_c1  antiRepFlank   \n",
       "\n",
       "                                                 seq  length  length_ctg  \\\n",
       "0  ATAACTCTTCCACATACTTCAGAATTTTATTTAGATGCTGAAGTTA...     350       17162   \n",
       "0  TGTACTTTTAATATTTATATCATTATTGAACGTTTCGTAATTCAAT...     350       17162   \n",
       "1  ACACACTTCATAAGGAGTATGTGAATAATCCCATATCTCTTGAAAC...     350       45775   \n",
       "1  GTGGGCGTTATTAAAAAGATAAAAAAAGAGAAAAAATATAAGGCAT...     350       45775   \n",
       "2  TCTTTTAGAAATATTAATATTCATAACCTTACTCAATTTATCTAAT...     350       79404   \n",
       "\n",
       "                         oid  \n",
       "0  GCA_001594005.1_m_1_c1_a4  \n",
       "0  GCA_001594005.1_m_1_c1_a4  \n",
       "1  GCA_001594005.1_m_1_c1_a5  \n",
       "1  GCA_001594005.1_m_1_c1_a5  \n",
       "2  GCA_001594005.1_m_2_c1_a3  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Anti-repeat flanking regions calc\n",
    "print(\">\", \"Anti-repeat flanking regions calculation\")\n",
    "atfRes = flankReg(antRes_ftd, mode=\"<.>\", length=350, genomeDir=asm_o_dir, component=\"antiRepFlank\")\n",
    "#validateSeq(atfRes)\n",
    "print(atfRes.shape)\n",
    "display(atfRes.head())\n",
    "# write\n",
    "atfRes.to_csv(all_dir + \"/antiRepFlank.txt\", header=True, sep=\"\\t\", quoting=csv.QUOTE_NONE, index=False)\n",
    "recs = [SeqRecord.SeqRecord(Seq.Seq(row[\"seq\"]), id=row[\"id\"], description=\"\") for i, row in atfRes.iterrows()]\n",
    "_ = SeqIO.write(recs, all_dir + \"/antiRepFlank.fa\", \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "66b50911-6a85-4ed4-be58-86aae2e28da9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> Terminator identification\n"
     ]
    }
   ],
   "source": [
    "# Terminator identification\n",
    "print(\">\", \"Terminator identification\")\n",
    "tmn_out = all_dir + \"/terminator_erpin.out\"\n",
    "tmn_err = all_dir + \"/terminator_erpin.err\"\n",
    "if not os.path.exists(all_dir + \"/terminator_erpin.out\"):\n",
    "    ret = os.system(\"scripts/Arnold/erpin scripts/Arnold/rho-indep.epn {} -1,4 -add 2 4 2 -pcw 3.0 -cutoff 100% > {} 2>{}\".format(all_dir + \"/antiRepFlank.fa\", tmn_out, tmn_err))\n",
    "    if ret != 0:\n",
    "        print(\"Please confirm that erpin exists.\", file=sys.stderr)\n",
    "else:\n",
    "    print(\"{} exists, skipped.\".format(tmn_out))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "471e9945-28ec-472f-81f6-8db779345643",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> Terminator res extraction\n",
      ">> Before filtering: 8539 / 20092 anti-repeat flanks contain terminators.\n",
      "(13052, 8)\n",
      ">> After filtering: 5247 / 20092 anti-repeat flanks contain terminators.\n",
      "(6543, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>atf_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>seq</th>\n",
       "      <th>score</th>\n",
       "      <th>evalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_001594005.1_m_1_c1_a4_l_f1</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a4_l</td>\n",
       "      <td>216</td>\n",
       "      <td>246</td>\n",
       "      <td>+</td>\n",
       "      <td>GCCAGTAA.gt----------------------------.ATGCTG...</td>\n",
       "      <td>9.45</td>\n",
       "      <td>1.37e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_001594005.1_m_1_c1_a4_l_f2</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a4_l</td>\n",
       "      <td>249</td>\n",
       "      <td>302</td>\n",
       "      <td>+</td>\n",
       "      <td>AAAACATT.tgcttgaattatttcccaggatat------.ATTGTT...</td>\n",
       "      <td>9.90</td>\n",
       "      <td>1.05e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_001594005.1_m_1_c1_a4_r_f1</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a4_r</td>\n",
       "      <td>31</td>\n",
       "      <td>78</td>\n",
       "      <td>+</td>\n",
       "      <td>CGTTTCGT.aattcaatactattttac------------.CTGAAA...</td>\n",
       "      <td>9.96</td>\n",
       "      <td>1.02e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GCA_001594005.1_m_1_c1_a5_l_f1</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a5_l</td>\n",
       "      <td>12</td>\n",
       "      <td>64</td>\n",
       "      <td>+</td>\n",
       "      <td>AAGGAGTA.tgtgaataatcccatatctcttga------.AACTCT...</td>\n",
       "      <td>8.40</td>\n",
       "      <td>2.43e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GCA_001594005.1_m_1_c1_a5_r_f1</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a5_r</td>\n",
       "      <td>63</td>\n",
       "      <td>117</td>\n",
       "      <td>+</td>\n",
       "      <td>AAGTCATG.tgatttgggcaaaaagcttgaattta----.TATGGC...</td>\n",
       "      <td>8.30</td>\n",
       "      <td>2.56e+03</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               id                       atf_id  start  end  \\\n",
       "0  GCA_001594005.1_m_1_c1_a4_l_f1  GCA_001594005.1_m_1_c1_a4_l    216  246   \n",
       "1  GCA_001594005.1_m_1_c1_a4_l_f2  GCA_001594005.1_m_1_c1_a4_l    249  302   \n",
       "3  GCA_001594005.1_m_1_c1_a4_r_f1  GCA_001594005.1_m_1_c1_a4_r     31   78   \n",
       "5  GCA_001594005.1_m_1_c1_a5_l_f1  GCA_001594005.1_m_1_c1_a5_l     12   64   \n",
       "7  GCA_001594005.1_m_1_c1_a5_r_f1  GCA_001594005.1_m_1_c1_a5_r     63  117   \n",
       "\n",
       "  strand                                                seq score    evalue  \n",
       "0      +  GCCAGTAA.gt----------------------------.ATGCTG...  9.45  1.37e+03  \n",
       "1      +  AAAACATT.tgcttgaattatttcccaggatat------.ATTGTT...  9.90  1.05e+03  \n",
       "3      +  CGTTTCGT.aattcaatactattttac------------.CTGAAA...  9.96  1.02e+03  \n",
       "5      +  AAGGAGTA.tgtgaataatcccatatctcttga------.AACTCT...  8.40  2.43e+03  \n",
       "7      +  AAGTCATG.tgatttgggcaaaaagcttgaattta----.TATGGC...  8.30  2.56e+03  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Terminator res extraction\n",
    "print(\">\", \"Terminator res extraction\")\n",
    "temRes = readERP(\"output/all/terminator_erpin.out\")\n",
    "print(\">> Before filtering: {} / {} anti-repeat flanks contain terminators.\".format(len(temRes[\"atf_id\"].unique()), len(atfRes[\"id\"].unique())))\n",
    "print(temRes.shape)\n",
    "temRes = temRes[temRes[\"strand\"] == \"+\"]\n",
    "print(\">> After filtering: {} / {} anti-repeat flanks contain terminators.\".format(len(temRes[\"atf_id\"].unique()), len(atfRes[\"id\"].unique())))\n",
    "print(temRes.shape)\n",
    "temRes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "cc6c9bb4-e70b-4c54-845e-897467078d2d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.247"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# stat\n",
    "stat_tem = pd.DataFrame(temRes[\"atf_id\"].value_counts())\n",
    "round(stat_tem[\"atf_id\"].mean(), 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "4d1220c2-2f82-4924-9abb-50b6e9091c5d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> Terminator global position calculation\n",
      "(6543, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>id</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>asm</th>\n",
       "      <th>tool</th>\n",
       "      <th>conid</th>\n",
       "      <th>component</th>\n",
       "      <th>seq</th>\n",
       "      <th>length</th>\n",
       "      <th>length_ctg</th>\n",
       "      <th>oid</th>\n",
       "      <th>atf_id</th>\n",
       "      <th>evalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LTBA01000045.1</td>\n",
       "      <td>15418</td>\n",
       "      <td>15449</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a4_l_f1</td>\n",
       "      <td>900</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>minCED</td>\n",
       "      <td>GCA_001594005.1_m_1_c1</td>\n",
       "      <td>teminator</td>\n",
       "      <td>GCCAGTAAGTATGCTGGCTTATTTTGCTAAA</td>\n",
       "      <td>31</td>\n",
       "      <td>17162</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a4</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a4_l</td>\n",
       "      <td>1.37e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LTBA01000045.1</td>\n",
       "      <td>15362</td>\n",
       "      <td>15416</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a4_l_f2</td>\n",
       "      <td>900</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>minCED</td>\n",
       "      <td>GCA_001594005.1_m_1_c1</td>\n",
       "      <td>teminator</td>\n",
       "      <td>AAAACATTTGCTTGAATTATTTCCCAGGATATATTGTTTTTTTCTT...</td>\n",
       "      <td>54</td>\n",
       "      <td>17162</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a4</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a4_l</td>\n",
       "      <td>1.05e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LTBA01000045.1</td>\n",
       "      <td>15709</td>\n",
       "      <td>15757</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a4_r_f1</td>\n",
       "      <td>900</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>minCED</td>\n",
       "      <td>GCA_001594005.1_m_1_c1</td>\n",
       "      <td>teminator</td>\n",
       "      <td>CGTTTCGTAATTCAATACTATTTTACCTGAAACGCTTTTACCTTTTAT</td>\n",
       "      <td>48</td>\n",
       "      <td>17162</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a4</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a4_r</td>\n",
       "      <td>1.02e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LTBA01000021.1</td>\n",
       "      <td>28934</td>\n",
       "      <td>28987</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a5_l_f1</td>\n",
       "      <td>900</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>minCED</td>\n",
       "      <td>GCA_001594005.1_m_1_c1</td>\n",
       "      <td>teminator</td>\n",
       "      <td>AAGGAGTATGTGAATAATCCCATATCTCTTGAAACTCTTTTTGTTG...</td>\n",
       "      <td>53</td>\n",
       "      <td>45775</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a5</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a5_l</td>\n",
       "      <td>2.43e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LTBA01000021.1</td>\n",
       "      <td>29074</td>\n",
       "      <td>29129</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a5_r_f1</td>\n",
       "      <td>900</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>minCED</td>\n",
       "      <td>GCA_001594005.1_m_1_c1</td>\n",
       "      <td>teminator</td>\n",
       "      <td>AAGTCATGTGATTTGGGCAAAAAGCTTGAATTTATATGGCTTTTGT...</td>\n",
       "      <td>55</td>\n",
       "      <td>45775</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a5</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a5_r</td>\n",
       "      <td>2.56e+03</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           contig  start    end                              id  score strand  \\\n",
       "0  LTBA01000045.1  15418  15449  GCA_001594005.1_m_1_c1_a4_l_f1    900      -   \n",
       "1  LTBA01000045.1  15362  15416  GCA_001594005.1_m_1_c1_a4_l_f2    900      -   \n",
       "2  LTBA01000045.1  15709  15757  GCA_001594005.1_m_1_c1_a4_r_f1    900      +   \n",
       "3  LTBA01000021.1  28934  28987  GCA_001594005.1_m_1_c1_a5_l_f1    900      -   \n",
       "4  LTBA01000021.1  29074  29129  GCA_001594005.1_m_1_c1_a5_r_f1    900      +   \n",
       "\n",
       "               asm    tool                   conid  component  \\\n",
       "0  GCA_001594005.1  minCED  GCA_001594005.1_m_1_c1  teminator   \n",
       "1  GCA_001594005.1  minCED  GCA_001594005.1_m_1_c1  teminator   \n",
       "2  GCA_001594005.1  minCED  GCA_001594005.1_m_1_c1  teminator   \n",
       "3  GCA_001594005.1  minCED  GCA_001594005.1_m_1_c1  teminator   \n",
       "4  GCA_001594005.1  minCED  GCA_001594005.1_m_1_c1  teminator   \n",
       "\n",
       "                                                 seq  length  length_ctg  \\\n",
       "0                    GCCAGTAAGTATGCTGGCTTATTTTGCTAAA      31       17162   \n",
       "1  AAAACATTTGCTTGAATTATTTCCCAGGATATATTGTTTTTTTCTT...      54       17162   \n",
       "2   CGTTTCGTAATTCAATACTATTTTACCTGAAACGCTTTTACCTTTTAT      48       17162   \n",
       "3  AAGGAGTATGTGAATAATCCCATATCTCTTGAAACTCTTTTTGTTG...      53       45775   \n",
       "4  AAGTCATGTGATTTGGGCAAAAAGCTTGAATTTATATGGCTTTTGT...      55       45775   \n",
       "\n",
       "                         oid                       atf_id    evalue  \n",
       "0  GCA_001594005.1_m_1_c1_a4  GCA_001594005.1_m_1_c1_a4_l  1.37e+03  \n",
       "1  GCA_001594005.1_m_1_c1_a4  GCA_001594005.1_m_1_c1_a4_l  1.05e+03  \n",
       "2  GCA_001594005.1_m_1_c1_a4  GCA_001594005.1_m_1_c1_a4_r  1.02e+03  \n",
       "3  GCA_001594005.1_m_1_c1_a5  GCA_001594005.1_m_1_c1_a5_l  2.43e+03  \n",
       "4  GCA_001594005.1_m_1_c1_a5  GCA_001594005.1_m_1_c1_a5_r  2.56e+03  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Terminator global position calc\n",
    "print(\">\", \"Terminator global position calculation\")\n",
    "tegRes = calcGlo(atfRes, temRes, genomeDir=asm_o_dir)\n",
    "tegRes = tegRes[[i for i in tegRes.columns if not i.endswith(\"_tem\")]]\n",
    "tegRes.to_csv(all_dir + \"/terminator.txt\", header=True, sep=\"\\t\", quoting=csv.QUOTE_NONE, index=False)\n",
    "print(tegRes.shape)\n",
    "tegRes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "c1053397-9efc-4c58-acf0-4f7576eb2255",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> TracrRNA calculation\n",
      "(6543, 20)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>id</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>asm</th>\n",
       "      <th>tool</th>\n",
       "      <th>conid</th>\n",
       "      <th>component</th>\n",
       "      <th>seq</th>\n",
       "      <th>length</th>\n",
       "      <th>evalue</th>\n",
       "      <th>cid</th>\n",
       "      <th>asm_cri</th>\n",
       "      <th>contig_cri</th>\n",
       "      <th>start_cri</th>\n",
       "      <th>end_cri</th>\n",
       "      <th>length_cri</th>\n",
       "      <th>ctdist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LTBA01000045.1</td>\n",
       "      <td>15418</td>\n",
       "      <td>15679</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a4_l_f1_t1</td>\n",
       "      <td>900</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>minCED</td>\n",
       "      <td>GCA_001594005.1_m_1_c1</td>\n",
       "      <td>tracrRNA</td>\n",
       "      <td>TCAGGGGATATTAATATAACTCTTCCACATACTTCAGAATTTTATT...</td>\n",
       "      <td>261</td>\n",
       "      <td>1.37e+03</td>\n",
       "      <td>GCA_001594005.1_m_1</td>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>LTBA01000005.1</td>\n",
       "      <td>79637</td>\n",
       "      <td>79871</td>\n",
       "      <td>234</td>\n",
       "      <td>1000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LTBA01000045.1</td>\n",
       "      <td>15362</td>\n",
       "      <td>15679</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a4_l_f2_t1</td>\n",
       "      <td>900</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>minCED</td>\n",
       "      <td>GCA_001594005.1_m_1_c1</td>\n",
       "      <td>tracrRNA</td>\n",
       "      <td>TCAGGGGATATTAATATAACTCTTCCACATACTTCAGAATTTTATT...</td>\n",
       "      <td>317</td>\n",
       "      <td>1.05e+03</td>\n",
       "      <td>GCA_001594005.1_m_1</td>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>LTBA01000005.1</td>\n",
       "      <td>79637</td>\n",
       "      <td>79871</td>\n",
       "      <td>234</td>\n",
       "      <td>1000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LTBA01000045.1</td>\n",
       "      <td>15664</td>\n",
       "      <td>15757</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a4_r_f1_t1</td>\n",
       "      <td>900</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>minCED</td>\n",
       "      <td>GCA_001594005.1_m_1_c1</td>\n",
       "      <td>tracrRNA</td>\n",
       "      <td>ATTAATATCCCCTGATGTACTTTTAATATTTATATCATTATTGAAC...</td>\n",
       "      <td>93</td>\n",
       "      <td>1.02e+03</td>\n",
       "      <td>GCA_001594005.1_m_1</td>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>LTBA01000005.1</td>\n",
       "      <td>79637</td>\n",
       "      <td>79871</td>\n",
       "      <td>234</td>\n",
       "      <td>1000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LTBA01000021.1</td>\n",
       "      <td>28934</td>\n",
       "      <td>29012</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a5_l_f1_t1</td>\n",
       "      <td>900</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>minCED</td>\n",
       "      <td>GCA_001594005.1_m_1_c1</td>\n",
       "      <td>tracrRNA</td>\n",
       "      <td>TCACCATTAATATCACACACTTCATAAGGAGTATGTGAATAATCCC...</td>\n",
       "      <td>78</td>\n",
       "      <td>2.43e+03</td>\n",
       "      <td>GCA_001594005.1_m_1</td>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>LTBA01000005.1</td>\n",
       "      <td>79637</td>\n",
       "      <td>79871</td>\n",
       "      <td>234</td>\n",
       "      <td>1000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LTBA01000021.1</td>\n",
       "      <td>28998</td>\n",
       "      <td>29129</td>\n",
       "      <td>GCA_001594005.1_m_1_c1_a5_r_f1_t1</td>\n",
       "      <td>900</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>minCED</td>\n",
       "      <td>GCA_001594005.1_m_1_c1</td>\n",
       "      <td>tracrRNA</td>\n",
       "      <td>GATATTAATGGTGAGTGGGCGTTATTAAAAAGATAAAAAAAGAGAA...</td>\n",
       "      <td>131</td>\n",
       "      <td>2.56e+03</td>\n",
       "      <td>GCA_001594005.1_m_1</td>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>LTBA01000005.1</td>\n",
       "      <td>79637</td>\n",
       "      <td>79871</td>\n",
       "      <td>234</td>\n",
       "      <td>1000000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           contig  start    end                                 id  score  \\\n",
       "0  LTBA01000045.1  15418  15679  GCA_001594005.1_m_1_c1_a4_l_f1_t1    900   \n",
       "1  LTBA01000045.1  15362  15679  GCA_001594005.1_m_1_c1_a4_l_f2_t1    900   \n",
       "2  LTBA01000045.1  15664  15757  GCA_001594005.1_m_1_c1_a4_r_f1_t1    900   \n",
       "3  LTBA01000021.1  28934  29012  GCA_001594005.1_m_1_c1_a5_l_f1_t1    900   \n",
       "4  LTBA01000021.1  28998  29129  GCA_001594005.1_m_1_c1_a5_r_f1_t1    900   \n",
       "\n",
       "  strand              asm    tool                   conid component  \\\n",
       "0      -  GCA_001594005.1  minCED  GCA_001594005.1_m_1_c1  tracrRNA   \n",
       "1      -  GCA_001594005.1  minCED  GCA_001594005.1_m_1_c1  tracrRNA   \n",
       "2      +  GCA_001594005.1  minCED  GCA_001594005.1_m_1_c1  tracrRNA   \n",
       "3      -  GCA_001594005.1  minCED  GCA_001594005.1_m_1_c1  tracrRNA   \n",
       "4      +  GCA_001594005.1  minCED  GCA_001594005.1_m_1_c1  tracrRNA   \n",
       "\n",
       "                                                 seq  length    evalue  \\\n",
       "0  TCAGGGGATATTAATATAACTCTTCCACATACTTCAGAATTTTATT...     261  1.37e+03   \n",
       "1  TCAGGGGATATTAATATAACTCTTCCACATACTTCAGAATTTTATT...     317  1.05e+03   \n",
       "2  ATTAATATCCCCTGATGTACTTTTAATATTTATATCATTATTGAAC...      93  1.02e+03   \n",
       "3  TCACCATTAATATCACACACTTCATAAGGAGTATGTGAATAATCCC...      78  2.43e+03   \n",
       "4  GATATTAATGGTGAGTGGGCGTTATTAAAAAGATAAAAAAAGAGAA...     131  2.56e+03   \n",
       "\n",
       "                   cid          asm_cri      contig_cri  start_cri  end_cri  \\\n",
       "0  GCA_001594005.1_m_1  GCA_001594005.1  LTBA01000005.1      79637    79871   \n",
       "1  GCA_001594005.1_m_1  GCA_001594005.1  LTBA01000005.1      79637    79871   \n",
       "2  GCA_001594005.1_m_1  GCA_001594005.1  LTBA01000005.1      79637    79871   \n",
       "3  GCA_001594005.1_m_1  GCA_001594005.1  LTBA01000005.1      79637    79871   \n",
       "4  GCA_001594005.1_m_1  GCA_001594005.1  LTBA01000005.1      79637    79871   \n",
       "\n",
       "   length_cri      ctdist  \n",
       "0         234  1000000000  \n",
       "1         234  1000000000  \n",
       "2         234  1000000000  \n",
       "3         234  1000000000  \n",
       "4         234  1000000000  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TracrRNA calculation\n",
    "print(\">\", \"TracrRNA calculation\")\n",
    "traRes = unionReg(antRes_ftd, tegRes, genomeDir=asm_o_dir, component=\"tracrRNA\")\n",
    "traRes = traRes[[i for i in traRes.columns if not i.endswith(\"_teg\")]].drop(columns=[\"oid\", \"atf_id\", \"length_ctg\"])\n",
    "traRes = pd.merge(traRes, conRes[[\"id\", \"cid\"]], left_on=\"conid\", right_on=\"id\", suffixes=[\"\", \"_con\"]).drop(columns=\"id_con\")\n",
    "traRes = pd.merge(traRes, criPos, on=\"cid\", suffixes=[\"\", \"_cri\"])\n",
    "# distance between tracrRNA and corresponding CRISPR array\n",
    "cads = []\n",
    "for _, row in traRes.iterrows():\n",
    "    if row[\"contig\"] == row[\"contig_cri\"]:\n",
    "        if row[\"start\"] >= row[\"end_cri\"]:\n",
    "            cad = row[\"start\"] - row[\"end_cri\"]\n",
    "        elif row[\"end\"] <= row[\"start_cri\"]:\n",
    "            cad = - (row[\"start_cri\"] - row[\"end\"])\n",
    "        else:\n",
    "            cad = 0\n",
    "    else:\n",
    "        cad = int(1e9)\n",
    "    cads.append(cad)\n",
    "traRes[\"ctdist\"] = cads\n",
    "traRes.to_csv(all_dir + \"/tracrRNA.txt\", header=True, sep=\"\\t\", quoting=csv.QUOTE_NONE, index=False)\n",
    "print(traRes.shape)\n",
    "traRes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "49248c30-3abc-4659-b072-eda0cf2df78e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> TracrRNA Filtering 1\n",
      ">> Filtering (not contain N)\n",
      "(6521, 20)\n",
      ">> Filtering (not contain TTTT)\n",
      "(1188, 20)\n"
     ]
    }
   ],
   "source": [
    "# TracrRNA Filtering 1\n",
    "print(\">\", \"TracrRNA Filtering 1\")\n",
    "print(\">>\", \"Filtering (not contain N)\")\n",
    "traRes = traRes[~ traRes[\"seq\"].str.contains(\"N\")]\n",
    "print(traRes.shape)\n",
    "print(\">>\", \"Filtering (not contain TTTT)\")\n",
    "traRes = traRes[~ traRes[\"seq\"].str.contains(\"TTTT\")]\n",
    "print(traRes.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "3e50cc0b-8173-4a3b-a8cf-9a648d898882",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> TracrRNA Filtering 2\n",
      "(1179, 20)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>id</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>asm</th>\n",
       "      <th>tool</th>\n",
       "      <th>conid</th>\n",
       "      <th>component</th>\n",
       "      <th>seq</th>\n",
       "      <th>length</th>\n",
       "      <th>evalue</th>\n",
       "      <th>cid</th>\n",
       "      <th>asm_cri</th>\n",
       "      <th>contig_cri</th>\n",
       "      <th>start_cri</th>\n",
       "      <th>end_cri</th>\n",
       "      <th>length_cri</th>\n",
       "      <th>ctdist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LTBA01000008.1</td>\n",
       "      <td>88113</td>\n",
       "      <td>88272</td>\n",
       "      <td>GCA_001594005.1_p_1_c1_a4_r_f1_t1</td>\n",
       "      <td>900</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>pilerCR</td>\n",
       "      <td>GCA_001594005.1_p_1_c1</td>\n",
       "      <td>tracrRNA</td>\n",
       "      <td>TTTCCATTAATATCTATATAACCTTGAGTATTACCATTGATTCTAA...</td>\n",
       "      <td>159</td>\n",
       "      <td>475.0</td>\n",
       "      <td>GCA_001594005.1_p_1</td>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>LTBA01000005.1</td>\n",
       "      <td>79736</td>\n",
       "      <td>79870</td>\n",
       "      <td>134</td>\n",
       "      <td>1000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RGDI01000306.1</td>\n",
       "      <td>1492</td>\n",
       "      <td>1563</td>\n",
       "      <td>GCA_009918665.1_m_14_c1_a8_l_f1_t1</td>\n",
       "      <td>900</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_009918665.1</td>\n",
       "      <td>minCED</td>\n",
       "      <td>GCA_009918665.1_m_14_c1</td>\n",
       "      <td>tracrRNA</td>\n",
       "      <td>CTGGGACTTCTGTTTCTCTCTCAGAGAAGATAACCCCGTATTTACT...</td>\n",
       "      <td>71</td>\n",
       "      <td>6550.0</td>\n",
       "      <td>GCA_009918665.1_m_14</td>\n",
       "      <td>GCA_009918665.1</td>\n",
       "      <td>RGDI01000363.1</td>\n",
       "      <td>229</td>\n",
       "      <td>377</td>\n",
       "      <td>148</td>\n",
       "      <td>1000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RGDI01000306.1</td>\n",
       "      <td>1551</td>\n",
       "      <td>1702</td>\n",
       "      <td>GCA_009918665.1_m_14_c1_a8_r_f1_t1</td>\n",
       "      <td>900</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_009918665.1</td>\n",
       "      <td>minCED</td>\n",
       "      <td>GCA_009918665.1_m_14_c1</td>\n",
       "      <td>tracrRNA</td>\n",
       "      <td>CAGAAGTCCCAGGTACCCGTAATACCTCTGTTGTCATTCATTCCTT...</td>\n",
       "      <td>151</td>\n",
       "      <td>1760.0</td>\n",
       "      <td>GCA_009918665.1_m_14</td>\n",
       "      <td>GCA_009918665.1</td>\n",
       "      <td>RGDI01000363.1</td>\n",
       "      <td>229</td>\n",
       "      <td>377</td>\n",
       "      <td>148</td>\n",
       "      <td>1000000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RGDI01000111.1</td>\n",
       "      <td>4905</td>\n",
       "      <td>5138</td>\n",
       "      <td>GCA_009918665.1_p_1_c1_a2_r_f1_t1</td>\n",
       "      <td>900</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_009918665.1</td>\n",
       "      <td>pilerCR</td>\n",
       "      <td>GCA_009918665.1_p_1_c1</td>\n",
       "      <td>tracrRNA</td>\n",
       "      <td>GCTTGATATCCAACTGATACCGCATATTGTCCTTGTGTCGATGCTC...</td>\n",
       "      <td>233</td>\n",
       "      <td>9020.0</td>\n",
       "      <td>GCA_009918665.1_p_1</td>\n",
       "      <td>GCA_009918665.1</td>\n",
       "      <td>RGDI01000111.1</td>\n",
       "      <td>1964</td>\n",
       "      <td>2092</td>\n",
       "      <td>128</td>\n",
       "      <td>2813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RGDD01000801.1</td>\n",
       "      <td>6</td>\n",
       "      <td>126</td>\n",
       "      <td>GCA_009918765.1_m_13_c1_a2_l_f1_t1</td>\n",
       "      <td>900</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_009918765.1</td>\n",
       "      <td>minCED</td>\n",
       "      <td>GCA_009918765.1_m_13_c1</td>\n",
       "      <td>tracrRNA</td>\n",
       "      <td>CCTTGTGGACCTTGAGGTCCAGAAACACCTTGCGGTCCACTTGGTC...</td>\n",
       "      <td>120</td>\n",
       "      <td>4330.0</td>\n",
       "      <td>GCA_009918765.1_m_13</td>\n",
       "      <td>GCA_009918765.1</td>\n",
       "      <td>RGDD01000801.1</td>\n",
       "      <td>585</td>\n",
       "      <td>749</td>\n",
       "      <td>164</td>\n",
       "      <td>-459</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           contig  start    end                                  id  score  \\\n",
       "0  LTBA01000008.1  88113  88272   GCA_001594005.1_p_1_c1_a4_r_f1_t1    900   \n",
       "1  RGDI01000306.1   1492   1563  GCA_009918665.1_m_14_c1_a8_l_f1_t1    900   \n",
       "2  RGDI01000306.1   1551   1702  GCA_009918665.1_m_14_c1_a8_r_f1_t1    900   \n",
       "3  RGDI01000111.1   4905   5138   GCA_009918665.1_p_1_c1_a2_r_f1_t1    900   \n",
       "4  RGDD01000801.1      6    126  GCA_009918765.1_m_13_c1_a2_l_f1_t1    900   \n",
       "\n",
       "  strand              asm     tool                    conid component  \\\n",
       "0      +  GCA_001594005.1  pilerCR   GCA_001594005.1_p_1_c1  tracrRNA   \n",
       "1      -  GCA_009918665.1   minCED  GCA_009918665.1_m_14_c1  tracrRNA   \n",
       "2      +  GCA_009918665.1   minCED  GCA_009918665.1_m_14_c1  tracrRNA   \n",
       "3      +  GCA_009918665.1  pilerCR   GCA_009918665.1_p_1_c1  tracrRNA   \n",
       "4      -  GCA_009918765.1   minCED  GCA_009918765.1_m_13_c1  tracrRNA   \n",
       "\n",
       "                                                 seq  length  evalue  \\\n",
       "0  TTTCCATTAATATCTATATAACCTTGAGTATTACCATTGATTCTAA...     159   475.0   \n",
       "1  CTGGGACTTCTGTTTCTCTCTCAGAGAAGATAACCCCGTATTTACT...      71  6550.0   \n",
       "2  CAGAAGTCCCAGGTACCCGTAATACCTCTGTTGTCATTCATTCCTT...     151  1760.0   \n",
       "3  GCTTGATATCCAACTGATACCGCATATTGTCCTTGTGTCGATGCTC...     233  9020.0   \n",
       "4  CCTTGTGGACCTTGAGGTCCAGAAACACCTTGCGGTCCACTTGGTC...     120  4330.0   \n",
       "\n",
       "                    cid          asm_cri      contig_cri  start_cri  end_cri  \\\n",
       "0   GCA_001594005.1_p_1  GCA_001594005.1  LTBA01000005.1      79736    79870   \n",
       "1  GCA_009918665.1_m_14  GCA_009918665.1  RGDI01000363.1        229      377   \n",
       "2  GCA_009918665.1_m_14  GCA_009918665.1  RGDI01000363.1        229      377   \n",
       "3   GCA_009918665.1_p_1  GCA_009918665.1  RGDI01000111.1       1964     2092   \n",
       "4  GCA_009918765.1_m_13  GCA_009918765.1  RGDD01000801.1        585      749   \n",
       "\n",
       "   length_cri      ctdist  \n",
       "0         134  1000000000  \n",
       "1         148  1000000000  \n",
       "2         148  1000000000  \n",
       "3         128        2813  \n",
       "4         164        -459  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TracrRNA Filtering 2\n",
    "print(\">\", \"TracrRNA Filtering 2\")\n",
    "os.system(\"/bin/bash -c \\\"bedtools subtract -A -a <(tail -n +2 {}) -b <(tail -n +2 {}) > {}\\\"\".format(all_dir + \"/tracrRNA.txt\", all_dir + \"/criSlp.txt\", all_dir + \"/tracrRNA_ftd.txt\"))\n",
    "traRes_ftd = pd.read_csv(all_dir + \"/tracrRNA_ftd.txt\", header=None, sep=\"\\t\", quoting=csv.QUOTE_NONE)\n",
    "traRes_ftd.columns = traRes.columns\n",
    "traRes_ftd.to_csv(all_dir + \"/tracrRNA_ftd.txt\", header=True, sep=\"\\t\", quoting=csv.QUOTE_NONE, index=False)\n",
    "recs = [SeqRecord.SeqRecord(Seq.Seq(row[\"seq\"]), id=row[\"id\"], description=\"\") for i, row in traRes_ftd.iterrows()]\n",
    "SeqIO.write(recs, all_dir + \"/tracrRNA_ftd.fa\", \"fasta\")\n",
    "print(traRes_ftd.shape)\n",
    "traRes_ftd.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "adcea4b3-b917-4045-b95f-9aeb441d9521",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>tool</th>\n",
       "      <th>asm</th>\n",
       "      <th>minCED</th>\n",
       "      <th>pilerCR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_001594005.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_009918665.1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_009918765.1</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_010672215.1</td>\n",
       "      <td>276.0</td>\n",
       "      <td>95.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCA_903819865.1</td>\n",
       "      <td>183.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GCA_903830605.1</td>\n",
       "      <td>169.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GCA_903838155.1</td>\n",
       "      <td>79.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GCA_903853905.1</td>\n",
       "      <td>105.0</td>\n",
       "      <td>38.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GCA_903901705.1</td>\n",
       "      <td>108.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GCA_903907625.1</td>\n",
       "      <td>112.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "tool              asm  minCED  pilerCR\n",
       "0     GCA_001594005.1     NaN      1.0\n",
       "1     GCA_009918665.1     2.0      1.0\n",
       "2     GCA_009918765.1     9.0      1.0\n",
       "3     GCA_010672215.1   276.0     95.0\n",
       "4     GCA_903819865.1   183.0      NaN\n",
       "5     GCA_903830605.1   169.0      NaN\n",
       "6     GCA_903838155.1    79.0      NaN\n",
       "7     GCA_903853905.1   105.0     38.0\n",
       "8     GCA_903901705.1   108.0      NaN\n",
       "9     GCA_903907625.1   112.0      NaN"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# stat\n",
    "traRes_ftd.pivot_table(index=\"asm\", columns=\"tool\", values=\"id\", aggfunc=lambda x: len(x)).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "82a70f18-9d1a-4bf2-932e-cba90062aa3e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>id</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>asm</th>\n",
       "      <th>tool</th>\n",
       "      <th>conid</th>\n",
       "      <th>component</th>\n",
       "      <th>seq</th>\n",
       "      <th>length</th>\n",
       "      <th>evalue</th>\n",
       "      <th>cid</th>\n",
       "      <th>asm_cri</th>\n",
       "      <th>contig_cri</th>\n",
       "      <th>start_cri</th>\n",
       "      <th>end_cri</th>\n",
       "      <th>length_cri</th>\n",
       "      <th>ctdist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RGDI01000111.1</td>\n",
       "      <td>4905</td>\n",
       "      <td>5138</td>\n",
       "      <td>GCA_009918665.1_p_1_c1_a2_r_f1_t1</td>\n",
       "      <td>900</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_009918665.1</td>\n",
       "      <td>pilerCR</td>\n",
       "      <td>GCA_009918665.1_p_1_c1</td>\n",
       "      <td>tracrRNA</td>\n",
       "      <td>GCTTGATATCCAACTGATACCGCATATTGTCCTTGTGTCGATGCTC...</td>\n",
       "      <td>233</td>\n",
       "      <td>9020.0</td>\n",
       "      <td>GCA_009918665.1_p_1</td>\n",
       "      <td>GCA_009918665.1</td>\n",
       "      <td>RGDI01000111.1</td>\n",
       "      <td>1964</td>\n",
       "      <td>2092</td>\n",
       "      <td>128</td>\n",
       "      <td>2813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RGDD01000801.1</td>\n",
       "      <td>6</td>\n",
       "      <td>126</td>\n",
       "      <td>GCA_009918765.1_m_13_c1_a2_l_f1_t1</td>\n",
       "      <td>900</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_009918765.1</td>\n",
       "      <td>minCED</td>\n",
       "      <td>GCA_009918765.1_m_13_c1</td>\n",
       "      <td>tracrRNA</td>\n",
       "      <td>CCTTGTGGACCTTGAGGTCCAGAAACACCTTGCGGTCCACTTGGTC...</td>\n",
       "      <td>120</td>\n",
       "      <td>4330.0</td>\n",
       "      <td>GCA_009918765.1_m_13</td>\n",
       "      <td>GCA_009918765.1</td>\n",
       "      <td>RGDD01000801.1</td>\n",
       "      <td>585</td>\n",
       "      <td>749</td>\n",
       "      <td>164</td>\n",
       "      <td>-459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>RGDD01000801.1</td>\n",
       "      <td>6</td>\n",
       "      <td>135</td>\n",
       "      <td>GCA_009918765.1_m_13_c1_a5_l_f1_t1</td>\n",
       "      <td>900</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_009918765.1</td>\n",
       "      <td>minCED</td>\n",
       "      <td>GCA_009918765.1_m_13_c1</td>\n",
       "      <td>tracrRNA</td>\n",
       "      <td>CCTTGTGGACCTTGTGGACCTTGAGGTCCAGAAACACCTTGCGGTC...</td>\n",
       "      <td>129</td>\n",
       "      <td>4330.0</td>\n",
       "      <td>GCA_009918765.1_m_13</td>\n",
       "      <td>GCA_009918765.1</td>\n",
       "      <td>RGDD01000801.1</td>\n",
       "      <td>585</td>\n",
       "      <td>749</td>\n",
       "      <td>164</td>\n",
       "      <td>-450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>RGDD01000801.1</td>\n",
       "      <td>6</td>\n",
       "      <td>159</td>\n",
       "      <td>GCA_009918765.1_m_13_c1_a6_l_f1_t1</td>\n",
       "      <td>900</td>\n",
       "      <td>-</td>\n",
       "      <td>GCA_009918765.1</td>\n",
       "      <td>minCED</td>\n",
       "      <td>GCA_009918765.1_m_13_c1</td>\n",
       "      <td>tracrRNA</td>\n",
       "      <td>TGTGGACCTTGAGGACCAGCAATACCTTGTGGACCTTGTGGACCTT...</td>\n",
       "      <td>153</td>\n",
       "      <td>4330.0</td>\n",
       "      <td>GCA_009918765.1_m_13</td>\n",
       "      <td>GCA_009918765.1</td>\n",
       "      <td>RGDD01000801.1</td>\n",
       "      <td>585</td>\n",
       "      <td>749</td>\n",
       "      <td>164</td>\n",
       "      <td>-426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>RGDD01000334.1</td>\n",
       "      <td>1850</td>\n",
       "      <td>2128</td>\n",
       "      <td>GCA_009918765.1_p_1_c1_a2_r_f1_t1</td>\n",
       "      <td>900</td>\n",
       "      <td>+</td>\n",
       "      <td>GCA_009918765.1</td>\n",
       "      <td>pilerCR</td>\n",
       "      <td>GCA_009918765.1_p_1_c1</td>\n",
       "      <td>tracrRNA</td>\n",
       "      <td>GCAGTTGACCCTTGTCTGGGCTGCTGCAAGCTCAGTACAAGTGTCA...</td>\n",
       "      <td>278</td>\n",
       "      <td>9210.0</td>\n",
       "      <td>GCA_009918765.1_p_1</td>\n",
       "      <td>GCA_009918765.1</td>\n",
       "      <td>RGDD01000334.1</td>\n",
       "      <td>805</td>\n",
       "      <td>986</td>\n",
       "      <td>181</td>\n",
       "      <td>864</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            contig  start   end                                  id  score  \\\n",
       "3   RGDI01000111.1   4905  5138   GCA_009918665.1_p_1_c1_a2_r_f1_t1    900   \n",
       "4   RGDD01000801.1      6   126  GCA_009918765.1_m_13_c1_a2_l_f1_t1    900   \n",
       "5   RGDD01000801.1      6   135  GCA_009918765.1_m_13_c1_a5_l_f1_t1    900   \n",
       "6   RGDD01000801.1      6   159  GCA_009918765.1_m_13_c1_a6_l_f1_t1    900   \n",
       "13  RGDD01000334.1   1850  2128   GCA_009918765.1_p_1_c1_a2_r_f1_t1    900   \n",
       "\n",
       "   strand              asm     tool                    conid component  \\\n",
       "3       +  GCA_009918665.1  pilerCR   GCA_009918665.1_p_1_c1  tracrRNA   \n",
       "4       -  GCA_009918765.1   minCED  GCA_009918765.1_m_13_c1  tracrRNA   \n",
       "5       -  GCA_009918765.1   minCED  GCA_009918765.1_m_13_c1  tracrRNA   \n",
       "6       -  GCA_009918765.1   minCED  GCA_009918765.1_m_13_c1  tracrRNA   \n",
       "13      +  GCA_009918765.1  pilerCR   GCA_009918765.1_p_1_c1  tracrRNA   \n",
       "\n",
       "                                                  seq  length  evalue  \\\n",
       "3   GCTTGATATCCAACTGATACCGCATATTGTCCTTGTGTCGATGCTC...     233  9020.0   \n",
       "4   CCTTGTGGACCTTGAGGTCCAGAAACACCTTGCGGTCCACTTGGTC...     120  4330.0   \n",
       "5   CCTTGTGGACCTTGTGGACCTTGAGGTCCAGAAACACCTTGCGGTC...     129  4330.0   \n",
       "6   TGTGGACCTTGAGGACCAGCAATACCTTGTGGACCTTGTGGACCTT...     153  4330.0   \n",
       "13  GCAGTTGACCCTTGTCTGGGCTGCTGCAAGCTCAGTACAAGTGTCA...     278  9210.0   \n",
       "\n",
       "                     cid          asm_cri      contig_cri  start_cri  end_cri  \\\n",
       "3    GCA_009918665.1_p_1  GCA_009918665.1  RGDI01000111.1       1964     2092   \n",
       "4   GCA_009918765.1_m_13  GCA_009918765.1  RGDD01000801.1        585      749   \n",
       "5   GCA_009918765.1_m_13  GCA_009918765.1  RGDD01000801.1        585      749   \n",
       "6   GCA_009918765.1_m_13  GCA_009918765.1  RGDD01000801.1        585      749   \n",
       "13   GCA_009918765.1_p_1  GCA_009918765.1  RGDD01000334.1        805      986   \n",
       "\n",
       "    length_cri  ctdist  \n",
       "3          128    2813  \n",
       "4          164    -459  \n",
       "5          164    -450  \n",
       "6          164    -426  \n",
       "13         181     864  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "traRes_ftd[traRes_ftd[\"ctdist\"].abs() < 1e6]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "ce19ed13-208a-4824-b6cd-22a17c9d6139",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> Stat for the whole process\n",
      "14\n",
      "12\n",
      "12\n",
      "10\n",
      "10\n",
      "10\n",
      "10\n"
     ]
    }
   ],
   "source": [
    "# Stat for the whole process\n",
    "print(\">\", \"Stat for the whole process\")\n",
    "print(len(asms))\n",
    "print(len(criRes[\"asm\"].unique()))\n",
    "print(len(antRes[\"asm\"].unique()))\n",
    "print(len(antRes_ftd[\"asm\"].unique()))\n",
    "print(len(tegRes[\"asm\"].unique()))\n",
    "print(len(traRes[\"asm\"].unique()))\n",
    "print(len(traRes_ftd[\"asm\"].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17c0d2b0-45d2-40be-a235-cb4b1c869b12",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
